<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Brings Seurat to the Tidyverse  • tidyseurat</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Brings Seurat to the Tidyverse ">
<meta property="og:description" content="It creates an invisible layer that allow to see the Seurat object 
    as tibble and interact seamlessly with the tidyverse.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">tidyseurat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.13</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/introduction.html">Overview of the tidyseurat package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div id="tidyseurat---part-of-tidytranscriptomics" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#tidyseurat---part-of-tidytranscriptomics" class="anchor"></a>tidyseurat - part of tidytranscriptomics</h1></div>
<!-- badges: start -->

<p><strong>Brings Seurat to the tidyverse!</strong></p>
<p>website: <a href="https://stemangiola.github.io/tidyseurat/">stemangiola.github.io/tidyseurat/</a></p>
<p>Please also have a look at</p>
<ul>
<li>
<a href="https://stemangiola.github.io/tidybulk/">tidybulk</a> for tidy bulk RNA-seq analysis</li>
<li>
<a href="https://github.com/stemangiola/nanny">nanny</a> for tidy high-level data analysis and manipulation</li>
<li>
<a href="https://github.com/stemangiola/tidygate">tidygate</a> for adding custom gate information to your tibble</li>
<li>
<a href="https://stemangiola.github.io/tidyHeatmap/">tidyHeatmap</a> for heatmaps produced with tidy principles</li>
</ul>
<p><img src="reference/figures/logo_interaction-01.png" alt="visual cue"></p>
</div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>tidyseurat provides a bridge between the Seurat single-cell package <br><span class="math display">@<em>b</em><em>u</em><em>t</em><em>l</em><em>e</em><em>r</em>2018<em>i</em><em>n</em><em>t</em><em>e</em><em>g</em><em>r</em><em>a</em><em>t</em><em>i</em><em>n</em><em>g</em>; @<em>s</em><em>t</em><em>u</em><em>a</em><em>r</em><em>t</em>2019<em>c</em><em>o</em><em>m</em><em>p</em><em>r</em><em>e</em><em>h</em><em>e</em><em>n</em><em>s</em><em>i</em><em>v</em><em>e</em></span><br> and the tidyverse <br><span class="math display">@<em>w</em><em>i</em><em>c</em><em>k</em><em>h</em><em>a</em><em>m</em>2019<em>w</em><em>e</em><em>l</em><em>c</em><em>o</em><em>m</em><em>e</em></span><br>. It creates an invisible layer that enables viewing the Seurat object as a tidyverse tibble, and provides Seurat-compatible <em>dplyr</em>, <em>tidyr</em>, <em>ggplot</em> and <em>plotly</em> functions.</p>
<div id="functionsutilities-available" class="section level2">
<h2 class="hasAnchor">
<a href="#functionsutilities-available" class="anchor"></a>Functions/utilities available</h2>
<table class="table">
<thead><tr class="header">
<th>Seurat-compatible Functions</th>
<th>Description</th>
</tr></thead>
<tbody><tr class="odd">
<td><code>all</code></td>
<td>After all <code>tidyseurat</code> is a Seurat object, just better</td>
</tr></tbody>
</table>
<table class="table">
<thead><tr class="header">
<th>tidyverse Packages</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>dplyr</code></td>
<td>All <code>dplyr</code> APIs like for any tibble</td>
</tr>
<tr class="even">
<td><code>tidyr</code></td>
<td>All <code>tidyr</code> APIs like for any tibble</td>
</tr>
<tr class="odd">
<td><code>ggplot2</code></td>
<td>
<code>ggplot</code> like for any tibble</td>
</tr>
<tr class="even">
<td><code>plotly</code></td>
<td>
<code>plot_ly</code> like for any tibble</td>
</tr>
</tbody>
</table>
<table class="table">
<thead><tr class="header">
<th>Utilities</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>tidy</code></td>
<td>Add <code>tidyseurat</code> invisible layer over a Seurat object</td>
</tr>
<tr class="even">
<td><code>as_tibble</code></td>
<td>Convert cell-wise information to a <code>tbl_df</code>
</td>
</tr>
<tr class="odd">
<td><code>join_transcripts</code></td>
<td>Add transcript-wise information, returns a <code>tbl_df</code>
</td>
</tr>
</tbody>
</table>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>From CRAN</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">install.packages</span>(<span class="st">"tidyseurat"</span>)</span></code></pre></div>
<p>From Github (development)</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">"stemangiola/tidyseurat"</span>)</span>
<span id="cb2-2"><a href="#cb2-2"></a></span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="kw">library</span>(dplyr)</span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="kw">library</span>(tidyr)</span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="kw">library</span>(purrr)</span>
<span id="cb2-6"><a href="#cb2-6"></a><span class="kw">library</span>(magrittr)</span>
<span id="cb2-7"><a href="#cb2-7"></a><span class="kw">library</span>(ggplot2)</span>
<span id="cb2-8"><a href="#cb2-8"></a><span class="kw">library</span>(Seurat)</span>
<span id="cb2-9"><a href="#cb2-9"></a><span class="kw">library</span>(tidyseurat)</span></code></pre></div>
</div>
<div id="create-tidyseurat-the-best-of-both-worlds" class="section level2">
<h2 class="hasAnchor">
<a href="#create-tidyseurat-the-best-of-both-worlds" class="anchor"></a>Create <code>tidyseurat</code>, the best of both worlds!</h2>
<p>This is a seurat object but it is evaluated as tibble. So it is fully compatible both with Seurat and tidyverse APIs.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>pbmc_small_tidy &lt;-<span class="st"> </span>tidyseurat<span class="op">::</span>pbmc_small <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">tidy</span>()</span></code></pre></div>
<p><strong>It looks like a tibble</strong></p>
<div class="sourceCode" id="cb4"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>pbmc_small_tidy</span>
<span id="cb4-2"><a href="#cb4-2"></a></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="co">## # A tibble: 80 x 16</span></span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="co">##    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups</span></span>
<span id="cb4-5"><a href="#cb4-5"></a><span class="co">##    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; </span></span>
<span id="cb4-6"><a href="#cb4-6"></a><span class="co">##  1 ATGC… SeuratPro…         70           47 0               A             g2    </span></span>
<span id="cb4-7"><a href="#cb4-7"></a><span class="co">##  2 CATG… SeuratPro…         85           52 0               A             g1    </span></span>
<span id="cb4-8"><a href="#cb4-8"></a><span class="co">##  3 GAAC… SeuratPro…         87           50 1               B             g2    </span></span>
<span id="cb4-9"><a href="#cb4-9"></a><span class="co">##  4 TGAC… SeuratPro…        127           56 0               A             g2    </span></span>
<span id="cb4-10"><a href="#cb4-10"></a><span class="co">##  5 AGTC… SeuratPro…        173           53 0               A             g2    </span></span>
<span id="cb4-11"><a href="#cb4-11"></a><span class="co">##  6 TCTG… SeuratPro…         70           48 0               A             g1    </span></span>
<span id="cb4-12"><a href="#cb4-12"></a><span class="co">##  7 TGGT… SeuratPro…         64           36 0               A             g1    </span></span>
<span id="cb4-13"><a href="#cb4-13"></a><span class="co">##  8 GCAG… SeuratPro…         72           45 0               A             g1    </span></span>
<span id="cb4-14"><a href="#cb4-14"></a><span class="co">##  9 GATA… SeuratPro…         52           36 0               A             g1    </span></span>
<span id="cb4-15"><a href="#cb4-15"></a><span class="co">## 10 AATG… SeuratPro…        100           41 0               A             g1    </span></span>
<span id="cb4-16"><a href="#cb4-16"></a><span class="co">## # … with 70 more rows, and 9 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,</span></span>
<span id="cb4-17"><a href="#cb4-17"></a><span class="co">## #   PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;,</span></span>
<span id="cb4-18"><a href="#cb4-18"></a><span class="co">## #   tSNE_2 &lt;dbl&gt;</span></span></code></pre></div>
<p><strong>But it is a Seurat object after all</strong></p>
<div class="sourceCode" id="cb5"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>pbmc_small_tidy<span class="op">@</span>assays</span>
<span id="cb5-2"><a href="#cb5-2"></a></span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="co">## $RNA</span></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="co">## Assay data with 230 features for 80 cells</span></span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="co">## Top 10 variable features:</span></span>
<span id="cb5-6"><a href="#cb5-6"></a><span class="co">##  PPBP, IGLL5, VDAC3, CD1C, AKR1C3, PF4, MYL9, GNLY, TREML1, CA2</span></span></code></pre></div>
</div>
</div>
<div id="annotation-polishing" class="section level1">
<h1 class="hasAnchor">
<a href="#annotation-polishing" class="anchor"></a>Annotation polishing</h1>
<p>We may have a column that contains the directory each run was taken from, such as the “file” column in <code>pbmc_small_tidy</code>.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>pbmc_small_tidy<span class="op">$</span>file[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span>
<span id="cb6-2"><a href="#cb6-2"></a></span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="co">##                                     ATGCCAGAACGACT </span></span>
<span id="cb6-4"><a href="#cb6-4"></a><span class="co">## "../data/sample2/outs/filtered_feature_bc_matrix/" </span></span>
<span id="cb6-5"><a href="#cb6-5"></a><span class="co">##                                     CATGGCCTGTGCAT </span></span>
<span id="cb6-6"><a href="#cb6-6"></a><span class="co">## "../data/sample1/outs/filtered_feature_bc_matrix/" </span></span>
<span id="cb6-7"><a href="#cb6-7"></a><span class="co">##                                     GAACCTGATGAACC </span></span>
<span id="cb6-8"><a href="#cb6-8"></a><span class="co">## "../data/sample2/outs/filtered_feature_bc_matrix/" </span></span>
<span id="cb6-9"><a href="#cb6-9"></a><span class="co">##                                     TGACTGGATTCTCA </span></span>
<span id="cb6-10"><a href="#cb6-10"></a><span class="co">## "../data/sample2/outs/filtered_feature_bc_matrix/" </span></span>
<span id="cb6-11"><a href="#cb6-11"></a><span class="co">##                                     AGTCAGACTGCACA </span></span>
<span id="cb6-12"><a href="#cb6-12"></a><span class="co">## "../data/sample2/outs/filtered_feature_bc_matrix/"</span></span></code></pre></div>
<p>We may want to extract the run/sample name out of it into a separate column. Tidyverse <code>extract</code> can be used to convert a character column into multiple columns using regular expression groups.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="co"># Create sample column</span></span>
<span id="cb7-2"><a href="#cb7-2"></a>pbmc_small_polished &lt;-</span>
<span id="cb7-3"><a href="#cb7-3"></a><span class="st">  </span>pbmc_small_tidy <span class="op">%&gt;%</span></span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="st">  </span><span class="kw">extract</span>(file, <span class="st">"sample"</span>, <span class="st">"../data/([a-z0-9]+)/outs.+"</span>, <span class="dt">remove =</span> <span class="ot">FALSE</span>)</span>
<span id="cb7-5"><a href="#cb7-5"></a><span class="co"># Reorder to have sample column up front</span></span>
<span id="cb7-6"><a href="#cb7-6"></a>pbmc_small_polished <span class="op">%&gt;%</span></span>
<span id="cb7-7"><a href="#cb7-7"></a><span class="st">  </span><span class="kw">select</span>(sample, <span class="kw">everything</span>())</span>
<span id="cb7-8"><a href="#cb7-8"></a></span>
<span id="cb7-9"><a href="#cb7-9"></a><span class="co">## # A tibble: 80 x 17</span></span>
<span id="cb7-10"><a href="#cb7-10"></a><span class="co">##    cell  sample orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents</span></span>
<span id="cb7-11"><a href="#cb7-11"></a><span class="co">##    &lt;chr&gt; &lt;chr&gt;  &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;        </span></span>
<span id="cb7-12"><a href="#cb7-12"></a><span class="co">##  1 ATGC… sampl… SeuratPro…         70           47 0               A            </span></span>
<span id="cb7-13"><a href="#cb7-13"></a><span class="co">##  2 CATG… sampl… SeuratPro…         85           52 0               A            </span></span>
<span id="cb7-14"><a href="#cb7-14"></a><span class="co">##  3 GAAC… sampl… SeuratPro…         87           50 1               B            </span></span>
<span id="cb7-15"><a href="#cb7-15"></a><span class="co">##  4 TGAC… sampl… SeuratPro…        127           56 0               A            </span></span>
<span id="cb7-16"><a href="#cb7-16"></a><span class="co">##  5 AGTC… sampl… SeuratPro…        173           53 0               A            </span></span>
<span id="cb7-17"><a href="#cb7-17"></a><span class="co">##  6 TCTG… sampl… SeuratPro…         70           48 0               A            </span></span>
<span id="cb7-18"><a href="#cb7-18"></a><span class="co">##  7 TGGT… sampl… SeuratPro…         64           36 0               A            </span></span>
<span id="cb7-19"><a href="#cb7-19"></a><span class="co">##  8 GCAG… sampl… SeuratPro…         72           45 0               A            </span></span>
<span id="cb7-20"><a href="#cb7-20"></a><span class="co">##  9 GATA… sampl… SeuratPro…         52           36 0               A            </span></span>
<span id="cb7-21"><a href="#cb7-21"></a><span class="co">## 10 AATG… sampl… SeuratPro…        100           41 0               A            </span></span>
<span id="cb7-22"><a href="#cb7-22"></a><span class="co">## # … with 70 more rows, and 10 more variables: groups &lt;chr&gt;,</span></span>
<span id="cb7-23"><a href="#cb7-23"></a><span class="co">## #   RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;,</span></span>
<span id="cb7-24"><a href="#cb7-24"></a><span class="co">## #   PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</span></span></code></pre></div>
</div>
<div id="preliminary-plots" class="section level1">
<h1 class="hasAnchor">
<a href="#preliminary-plots" class="anchor"></a>Preliminary plots</h1>
<p>Set colours and theme for plots.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="co"># Use colourblind-friendly colours</span></span>
<span id="cb8-2"><a href="#cb8-2"></a>friendly_cols &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"#88CCEE"</span>, <span class="st">"#CC6677"</span>, <span class="st">"#DDCC77"</span>, <span class="st">"#117733"</span>, <span class="st">"#332288"</span>, <span class="st">"#AA4499"</span>, <span class="st">"#44AA99"</span>, <span class="st">"#999933"</span>, <span class="st">"#882255"</span>, <span class="st">"#661100"</span>, <span class="st">"#6699CC"</span>)</span>
<span id="cb8-3"><a href="#cb8-3"></a></span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="co"># Set theme</span></span>
<span id="cb8-5"><a href="#cb8-5"></a>my_theme &lt;-</span>
<span id="cb8-6"><a href="#cb8-6"></a><span class="st">  </span><span class="kw">list</span>(</span>
<span id="cb8-7"><a href="#cb8-7"></a>    <span class="kw">scale_fill_manual</span>(<span class="dt">values =</span> friendly_cols),</span>
<span id="cb8-8"><a href="#cb8-8"></a>    <span class="kw">scale_color_manual</span>(<span class="dt">values =</span> friendly_cols),</span>
<span id="cb8-9"><a href="#cb8-9"></a>    <span class="kw">theme_bw</span>() <span class="op">+</span></span>
<span id="cb8-10"><a href="#cb8-10"></a><span class="st">      </span><span class="kw">theme</span>(</span>
<span id="cb8-11"><a href="#cb8-11"></a>        <span class="dt">panel.border =</span> <span class="kw">element_blank</span>(),</span>
<span id="cb8-12"><a href="#cb8-12"></a>        <span class="dt">axis.line =</span> <span class="kw">element_line</span>(),</span>
<span id="cb8-13"><a href="#cb8-13"></a>        <span class="dt">panel.grid.major =</span> <span class="kw">element_line</span>(<span class="dt">size =</span> <span class="fl">0.2</span>),</span>
<span id="cb8-14"><a href="#cb8-14"></a>        <span class="dt">panel.grid.minor =</span> <span class="kw">element_line</span>(<span class="dt">size =</span> <span class="fl">0.1</span>),</span>
<span id="cb8-15"><a href="#cb8-15"></a>        <span class="dt">text =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">12</span>),</span>
<span id="cb8-16"><a href="#cb8-16"></a>        <span class="dt">legend.position =</span> <span class="st">"bottom"</span>,</span>
<span id="cb8-17"><a href="#cb8-17"></a>        <span class="dt">aspect.ratio =</span> <span class="dv">1</span>,</span>
<span id="cb8-18"><a href="#cb8-18"></a>        <span class="dt">strip.background =</span> <span class="kw">element_blank</span>(),</span>
<span id="cb8-19"><a href="#cb8-19"></a>        <span class="dt">axis.title.x =</span> <span class="kw">element_text</span>(<span class="dt">margin =</span> <span class="kw">margin</span>(<span class="dt">t =</span> <span class="dv">10</span>, <span class="dt">r =</span> <span class="dv">10</span>, <span class="dt">b =</span> <span class="dv">10</span>, <span class="dt">l =</span> <span class="dv">10</span>)),</span>
<span id="cb8-20"><a href="#cb8-20"></a>        <span class="dt">axis.title.y =</span> <span class="kw">element_text</span>(<span class="dt">margin =</span> <span class="kw">margin</span>(<span class="dt">t =</span> <span class="dv">10</span>, <span class="dt">r =</span> <span class="dv">10</span>, <span class="dt">b =</span> <span class="dv">10</span>, <span class="dt">l =</span> <span class="dv">10</span>))</span>
<span id="cb8-21"><a href="#cb8-21"></a>      )</span>
<span id="cb8-22"><a href="#cb8-22"></a>  )</span></code></pre></div>
<p>We can treat <code>pbmc_small_polished</code> effectively as a normal tibble for plotting.</p>
<p>Here we plot number of transcripts per cell.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>pbmc_small_polished <span class="op">%&gt;%</span></span>
<span id="cb9-2"><a href="#cb9-2"></a><span class="st">  </span>tidyseurat<span class="op">::</span><span class="kw">ggplot</span>(<span class="kw">aes</span>(nFeature_RNA, <span class="dt">fill =</span> groups)) <span class="op">+</span></span>
<span id="cb9-3"><a href="#cb9-3"></a><span class="st">  </span><span class="kw">geom_histogram</span>() <span class="op">+</span></span>
<span id="cb9-4"><a href="#cb9-4"></a><span class="st">  </span>my_theme</span></code></pre></div>
<p><img src="reference/figures/plot1-1.png"><!-- --></p>
<p>Here we plot total transcripts per cell.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>pbmc_small_polished <span class="op">%&gt;%</span></span>
<span id="cb10-2"><a href="#cb10-2"></a><span class="st">  </span>tidyseurat<span class="op">::</span><span class="kw">ggplot</span>(<span class="kw">aes</span>(groups, nCount_RNA, <span class="dt">fill =</span> groups)) <span class="op">+</span></span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="st">  </span><span class="kw">geom_boxplot</span>(<span class="dt">outlier.shape =</span> <span class="ot">NA</span>) <span class="op">+</span></span>
<span id="cb10-4"><a href="#cb10-4"></a><span class="st">  </span><span class="kw">geom_jitter</span>(<span class="dt">width =</span> <span class="fl">0.1</span>) <span class="op">+</span></span>
<span id="cb10-5"><a href="#cb10-5"></a><span class="st">  </span>my_theme</span></code></pre></div>
<p><img src="reference/figures/plot2-1.png"><!-- --></p>
<p>Here we plot abundance of two transcripts for each group.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>pbmc_small_polished <span class="op">%&gt;%</span></span>
<span id="cb11-2"><a href="#cb11-2"></a><span class="st">  </span><span class="kw">join_transcripts</span>(<span class="dt">transcripts =</span> <span class="kw">c</span>(<span class="st">"HLA-DRA"</span>, <span class="st">"LYZ"</span>)) <span class="op">%&gt;%</span></span>
<span id="cb11-3"><a href="#cb11-3"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(groups, abundance_RNA <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, <span class="dt">fill =</span> groups)) <span class="op">+</span></span>
<span id="cb11-4"><a href="#cb11-4"></a><span class="st">  </span><span class="kw">geom_boxplot</span>(<span class="dt">outlier.shape =</span> <span class="ot">NA</span>) <span class="op">+</span></span>
<span id="cb11-5"><a href="#cb11-5"></a><span class="st">  </span><span class="kw">geom_jitter</span>(<span class="kw">aes</span>(<span class="dt">size =</span> nCount_RNA), <span class="dt">alpha =</span> <span class="fl">0.5</span>, <span class="dt">width =</span> <span class="fl">0.2</span>) <span class="op">+</span></span>
<span id="cb11-6"><a href="#cb11-6"></a><span class="st">  </span><span class="kw">scale_y_log10</span>() <span class="op">+</span></span>
<span id="cb11-7"><a href="#cb11-7"></a><span class="st">  </span>my_theme</span></code></pre></div>
<p><img src="reference/figures/unnamed-chunk-14-1.png"><!-- --></p>
</div>
<div id="preprocess-the-dataset" class="section level1">
<h1 class="hasAnchor">
<a href="#preprocess-the-dataset" class="anchor"></a>Preprocess the dataset</h1>
<p>Also you can treat the object as Seurat object and proceed with data processing.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>pbmc_small_pca &lt;-</span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="st">  </span>pbmc_small_polished <span class="op">%&gt;%</span></span>
<span id="cb12-3"><a href="#cb12-3"></a><span class="st">  </span><span class="kw">SCTransform</span>(<span class="dt">verbose =</span> <span class="ot">FALSE</span>) <span class="op">%&gt;%</span></span>
<span id="cb12-4"><a href="#cb12-4"></a><span class="st">  </span><span class="kw">FindVariableFeatures</span>(<span class="dt">verbose =</span> <span class="ot">FALSE</span>) <span class="op">%&gt;%</span></span>
<span id="cb12-5"><a href="#cb12-5"></a><span class="st">  </span><span class="kw">RunPCA</span>(<span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span>
<span id="cb12-6"><a href="#cb12-6"></a></span>
<span id="cb12-7"><a href="#cb12-7"></a>pbmc_small_pca</span>
<span id="cb12-8"><a href="#cb12-8"></a></span>
<span id="cb12-9"><a href="#cb12-9"></a><span class="co">## # A tibble: 80 x 19</span></span>
<span id="cb12-10"><a href="#cb12-10"></a><span class="co">##    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups</span></span>
<span id="cb12-11"><a href="#cb12-11"></a><span class="co">##    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; </span></span>
<span id="cb12-12"><a href="#cb12-12"></a><span class="co">##  1 ATGC… SeuratPro…         70           47 0               A             g2    </span></span>
<span id="cb12-13"><a href="#cb12-13"></a><span class="co">##  2 CATG… SeuratPro…         85           52 0               A             g1    </span></span>
<span id="cb12-14"><a href="#cb12-14"></a><span class="co">##  3 GAAC… SeuratPro…         87           50 1               B             g2    </span></span>
<span id="cb12-15"><a href="#cb12-15"></a><span class="co">##  4 TGAC… SeuratPro…        127           56 0               A             g2    </span></span>
<span id="cb12-16"><a href="#cb12-16"></a><span class="co">##  5 AGTC… SeuratPro…        173           53 0               A             g2    </span></span>
<span id="cb12-17"><a href="#cb12-17"></a><span class="co">##  6 TCTG… SeuratPro…         70           48 0               A             g1    </span></span>
<span id="cb12-18"><a href="#cb12-18"></a><span class="co">##  7 TGGT… SeuratPro…         64           36 0               A             g1    </span></span>
<span id="cb12-19"><a href="#cb12-19"></a><span class="co">##  8 GCAG… SeuratPro…         72           45 0               A             g1    </span></span>
<span id="cb12-20"><a href="#cb12-20"></a><span class="co">##  9 GATA… SeuratPro…         52           36 0               A             g1    </span></span>
<span id="cb12-21"><a href="#cb12-21"></a><span class="co">## 10 AATG… SeuratPro…        100           41 0               A             g1    </span></span>
<span id="cb12-22"><a href="#cb12-22"></a><span class="co">## # … with 70 more rows, and 12 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,</span></span>
<span id="cb12-23"><a href="#cb12-23"></a><span class="co">## #   sample &lt;chr&gt;, nCount_SCT &lt;dbl&gt;, nFeature_SCT &lt;int&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;,</span></span>
<span id="cb12-24"><a href="#cb12-24"></a><span class="co">## #   PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</span></span></code></pre></div>
<p>If a tool is not included in the tidyseurat collection, we can use <code>as_tibble</code> to permanently convert <code>tidyseurat</code> into tibble.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>pbmc_small_pca <span class="op">%&gt;%</span></span>
<span id="cb13-2"><a href="#cb13-2"></a><span class="st">  </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span></span>
<span id="cb13-3"><a href="#cb13-3"></a><span class="st">  </span><span class="kw">select</span>(<span class="kw">contains</span>(<span class="st">"PC"</span>), <span class="kw">everything</span>()) <span class="op">%&gt;%</span></span>
<span id="cb13-4"><a href="#cb13-4"></a><span class="st">  </span>GGally<span class="op">::</span><span class="kw">ggpairs</span>(<span class="dt">columns =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, ggplot2<span class="op">::</span><span class="kw">aes</span>(<span class="dt">colour =</span> groups)) <span class="op">+</span></span>
<span id="cb13-5"><a href="#cb13-5"></a><span class="st">  </span>my_theme</span></code></pre></div>
<p><img src="reference/figures/pc_plot-1.png"><!-- --></p>
</div>
<div id="identify-clusters" class="section level1">
<h1 class="hasAnchor">
<a href="#identify-clusters" class="anchor"></a>Identify clusters</h1>
<p>We proceed with cluster identification with Seurat.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>pbmc_small_cluster &lt;-</span>
<span id="cb14-2"><a href="#cb14-2"></a><span class="st">  </span>pbmc_small_pca <span class="op">%&gt;%</span></span>
<span id="cb14-3"><a href="#cb14-3"></a><span class="st">  </span><span class="kw">FindNeighbors</span>(<span class="dt">verbose =</span> <span class="ot">FALSE</span>) <span class="op">%&gt;%</span></span>
<span id="cb14-4"><a href="#cb14-4"></a><span class="st">  </span><span class="kw">FindClusters</span>(<span class="dt">method =</span> <span class="st">"igraph"</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span>
<span id="cb14-5"><a href="#cb14-5"></a></span>
<span id="cb14-6"><a href="#cb14-6"></a>pbmc_small_cluster</span>
<span id="cb14-7"><a href="#cb14-7"></a></span>
<span id="cb14-8"><a href="#cb14-8"></a><span class="co">## # A tibble: 80 x 21</span></span>
<span id="cb14-9"><a href="#cb14-9"></a><span class="co">##    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups</span></span>
<span id="cb14-10"><a href="#cb14-10"></a><span class="co">##    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; </span></span>
<span id="cb14-11"><a href="#cb14-11"></a><span class="co">##  1 ATGC… SeuratPro…         70           47 0               A             g2    </span></span>
<span id="cb14-12"><a href="#cb14-12"></a><span class="co">##  2 CATG… SeuratPro…         85           52 0               A             g1    </span></span>
<span id="cb14-13"><a href="#cb14-13"></a><span class="co">##  3 GAAC… SeuratPro…         87           50 1               B             g2    </span></span>
<span id="cb14-14"><a href="#cb14-14"></a><span class="co">##  4 TGAC… SeuratPro…        127           56 0               A             g2    </span></span>
<span id="cb14-15"><a href="#cb14-15"></a><span class="co">##  5 AGTC… SeuratPro…        173           53 0               A             g2    </span></span>
<span id="cb14-16"><a href="#cb14-16"></a><span class="co">##  6 TCTG… SeuratPro…         70           48 0               A             g1    </span></span>
<span id="cb14-17"><a href="#cb14-17"></a><span class="co">##  7 TGGT… SeuratPro…         64           36 0               A             g1    </span></span>
<span id="cb14-18"><a href="#cb14-18"></a><span class="co">##  8 GCAG… SeuratPro…         72           45 0               A             g1    </span></span>
<span id="cb14-19"><a href="#cb14-19"></a><span class="co">##  9 GATA… SeuratPro…         52           36 0               A             g1    </span></span>
<span id="cb14-20"><a href="#cb14-20"></a><span class="co">## 10 AATG… SeuratPro…        100           41 0               A             g1    </span></span>
<span id="cb14-21"><a href="#cb14-21"></a><span class="co">## # … with 70 more rows, and 14 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,</span></span>
<span id="cb14-22"><a href="#cb14-22"></a><span class="co">## #   sample &lt;chr&gt;, nCount_SCT &lt;dbl&gt;, nFeature_SCT &lt;int&gt;, SCT_snn_res.0.8 &lt;fct&gt;,</span></span>
<span id="cb14-23"><a href="#cb14-23"></a><span class="co">## #   seurat_clusters &lt;fct&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;,</span></span>
<span id="cb14-24"><a href="#cb14-24"></a><span class="co">## #   PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</span></span></code></pre></div>
<p>Now we can interrogate the object as if it was a regular tibble data frame.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>pbmc_small_cluster <span class="op">%&gt;%</span></span>
<span id="cb15-2"><a href="#cb15-2"></a><span class="st">  </span>tidyseurat<span class="op">::</span><span class="kw">count</span>(groups, seurat_clusters)</span>
<span id="cb15-3"><a href="#cb15-3"></a></span>
<span id="cb15-4"><a href="#cb15-4"></a><span class="co">## # A tibble: 8 x 3</span></span>
<span id="cb15-5"><a href="#cb15-5"></a><span class="co">##   groups seurat_clusters     n</span></span>
<span id="cb15-6"><a href="#cb15-6"></a><span class="co">##   &lt;chr&gt;  &lt;fct&gt;           &lt;int&gt;</span></span>
<span id="cb15-7"><a href="#cb15-7"></a><span class="co">## 1 g1     0                  17</span></span>
<span id="cb15-8"><a href="#cb15-8"></a><span class="co">## 2 g1     1                  14</span></span>
<span id="cb15-9"><a href="#cb15-9"></a><span class="co">## 3 g1     2                   9</span></span>
<span id="cb15-10"><a href="#cb15-10"></a><span class="co">## 4 g1     3                   4</span></span>
<span id="cb15-11"><a href="#cb15-11"></a><span class="co">## 5 g2     0                  13</span></span>
<span id="cb15-12"><a href="#cb15-12"></a><span class="co">## 6 g2     1                  12</span></span>
<span id="cb15-13"><a href="#cb15-13"></a><span class="co">## 7 g2     2                   6</span></span>
<span id="cb15-14"><a href="#cb15-14"></a><span class="co">## 8 g2     3                   5</span></span></code></pre></div>
<p>We can identify cluster markers using Seurat.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a><span class="co"># Identify top 10 markers per cluster</span></span>
<span id="cb16-2"><a href="#cb16-2"></a>markers &lt;-</span>
<span id="cb16-3"><a href="#cb16-3"></a><span class="st">  </span>pbmc_small_cluster <span class="op">%&gt;%</span></span>
<span id="cb16-4"><a href="#cb16-4"></a><span class="st">  </span><span class="kw">FindAllMarkers</span>(<span class="dt">only.pos =</span> <span class="ot">TRUE</span>, <span class="dt">min.pct =</span> <span class="fl">0.25</span>, <span class="dt">thresh.use =</span> <span class="fl">0.25</span>) <span class="op">%&gt;%</span></span>
<span id="cb16-5"><a href="#cb16-5"></a><span class="st">  </span><span class="kw">group_by</span>(cluster) <span class="op">%&gt;%</span></span>
<span id="cb16-6"><a href="#cb16-6"></a><span class="st">  </span><span class="kw">top_n</span>(<span class="dv">10</span>, avg_logFC)</span>
<span id="cb16-7"><a href="#cb16-7"></a></span>
<span id="cb16-8"><a href="#cb16-8"></a><span class="co"># Plot heatmap</span></span>
<span id="cb16-9"><a href="#cb16-9"></a>pbmc_small_cluster <span class="op">%&gt;%</span></span>
<span id="cb16-10"><a href="#cb16-10"></a><span class="st">  </span><span class="kw">DoHeatmap</span>(</span>
<span id="cb16-11"><a href="#cb16-11"></a>    <span class="dt">features =</span> markers<span class="op">$</span>gene,</span>
<span id="cb16-12"><a href="#cb16-12"></a>    <span class="dt">group.colors =</span> friendly_cols</span>
<span id="cb16-13"><a href="#cb16-13"></a>  )</span></code></pre></div>
<p><img src="reference/figures/unnamed-chunk-15-1.png"><!-- --></p>
</div>
<div id="reduce-dimensions" class="section level1">
<h1 class="hasAnchor">
<a href="#reduce-dimensions" class="anchor"></a>Reduce dimensions</h1>
<p>We can calculate the first 3 UMAP dimensions using the Seurat framework.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>pbmc_small_UMAP &lt;-</span>
<span id="cb17-2"><a href="#cb17-2"></a><span class="st">  </span>pbmc_small_cluster <span class="op">%&gt;%</span></span>
<span id="cb17-3"><a href="#cb17-3"></a><span class="st">  </span><span class="kw">RunUMAP</span>(<span class="dt">reduction =</span> <span class="st">"pca"</span>, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">15</span>, <span class="dt">n.components =</span> 3L, )</span></code></pre></div>
<p>And we can plot them using 3D plot using plotly.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a>pbmc_small_UMAP <span class="op">%&gt;%</span></span>
<span id="cb18-2"><a href="#cb18-2"></a><span class="st">  </span><span class="kw">plot_ly</span>(</span>
<span id="cb18-3"><a href="#cb18-3"></a>    <span class="dt">x =</span> <span class="op">~</span><span class="st">`</span><span class="dt">UMAP_1</span><span class="st">`</span>,</span>
<span id="cb18-4"><a href="#cb18-4"></a>    <span class="dt">y =</span> <span class="op">~</span><span class="st">`</span><span class="dt">UMAP_2</span><span class="st">`</span>,</span>
<span id="cb18-5"><a href="#cb18-5"></a>    <span class="dt">z =</span> <span class="op">~</span><span class="st">`</span><span class="dt">UMAP_3</span><span class="st">`</span>,</span>
<span id="cb18-6"><a href="#cb18-6"></a>    <span class="dt">color =</span> <span class="op">~</span>seurat_clusters,</span>
<span id="cb18-7"><a href="#cb18-7"></a>    <span class="dt">colors =</span> friendly_cols[<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>]</span>
<span id="cb18-8"><a href="#cb18-8"></a>  )</span></code></pre></div>
<p><img src="reference/figures/plotly.png" alt="screenshot plotly"></p>
<div id="cell-type-prediction" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-type-prediction" class="anchor"></a>Cell type prediction</h2>
<p>We can infer cell type identities using <em>SingleR</em> <br><span class="math display">@<em>a</em><em>r</em><em>a</em><em>n</em>2019<em>r</em><em>e</em><em>f</em><em>e</em><em>r</em><em>e</em><em>n</em><em>c</em><em>e</em></span><br> and manipulate the output using tidyverse.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a><span class="co"># Get cell type reference data</span></span>
<span id="cb19-2"><a href="#cb19-2"></a>blueprint &lt;-<span class="st"> </span>celldex<span class="op">::</span><span class="kw">BlueprintEncodeData</span>()</span>
<span id="cb19-3"><a href="#cb19-3"></a></span>
<span id="cb19-4"><a href="#cb19-4"></a><span class="co"># Infer cell identities</span></span>
<span id="cb19-5"><a href="#cb19-5"></a>cell_type_df &lt;-</span>
<span id="cb19-6"><a href="#cb19-6"></a><span class="st">  </span>pbmc_small_UMAP<span class="op">@</span>assays[[<span class="st">"SCT"</span>]]<span class="op">@</span>counts <span class="op">%&gt;%</span></span>
<span id="cb19-7"><a href="#cb19-7"></a><span class="st">  </span><span class="kw">log1p</span>() <span class="op">%&gt;%</span></span>
<span id="cb19-8"><a href="#cb19-8"></a><span class="st">  </span>Matrix<span class="op">::</span><span class="kw">Matrix</span>(<span class="dt">sparse =</span> <span class="ot">TRUE</span>) <span class="op">%&gt;%</span></span>
<span id="cb19-9"><a href="#cb19-9"></a><span class="st">  </span>SingleR<span class="op">::</span><span class="kw">SingleR</span>(</span>
<span id="cb19-10"><a href="#cb19-10"></a>    <span class="dt">ref =</span> blueprint,</span>
<span id="cb19-11"><a href="#cb19-11"></a>    <span class="dt">labels =</span> blueprint<span class="op">$</span>label.main,</span>
<span id="cb19-12"><a href="#cb19-12"></a>    <span class="dt">method =</span> <span class="st">"single"</span></span>
<span id="cb19-13"><a href="#cb19-13"></a>  ) <span class="op">%&gt;%</span></span>
<span id="cb19-14"><a href="#cb19-14"></a><span class="st">  </span><span class="kw">as.data.frame</span>() <span class="op">%&gt;%</span></span>
<span id="cb19-15"><a href="#cb19-15"></a><span class="st">  </span><span class="kw">as_tibble</span>(<span class="dt">rownames =</span> <span class="st">"cell"</span>) <span class="op">%&gt;%</span></span>
<span id="cb19-16"><a href="#cb19-16"></a><span class="st">  </span><span class="kw">select</span>(cell, first.labels)</span>
<span id="cb19-17"><a href="#cb19-17"></a></span>
<span id="cb19-18"><a href="#cb19-18"></a><span class="co"># Join UMAP and cell type info</span></span>
<span id="cb19-19"><a href="#cb19-19"></a>pbmc_small_cell_type &lt;-</span>
<span id="cb19-20"><a href="#cb19-20"></a><span class="st">  </span>pbmc_small_UMAP <span class="op">%&gt;%</span></span>
<span id="cb19-21"><a href="#cb19-21"></a><span class="st">  </span><span class="kw">left_join</span>(cell_type_df, <span class="dt">by =</span> <span class="st">"cell"</span>)</span>
<span id="cb19-22"><a href="#cb19-22"></a></span>
<span id="cb19-23"><a href="#cb19-23"></a><span class="co"># Reorder columns</span></span>
<span id="cb19-24"><a href="#cb19-24"></a>pbmc_small_cell_type <span class="op">%&gt;%</span></span>
<span id="cb19-25"><a href="#cb19-25"></a><span class="st">  </span>tidyseurat<span class="op">::</span><span class="kw">select</span>(cell, first.labels, <span class="kw">everything</span>())</span>
<span id="cb19-26"><a href="#cb19-26"></a></span>
<span id="cb19-27"><a href="#cb19-27"></a><span class="co">## # A tibble: 80 x 25</span></span>
<span id="cb19-28"><a href="#cb19-28"></a><span class="co">##    cell  first.labels orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8</span></span>
<span id="cb19-29"><a href="#cb19-29"></a><span class="co">##    &lt;chr&gt; &lt;chr&gt;        &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;          </span></span>
<span id="cb19-30"><a href="#cb19-30"></a><span class="co">##  1 ATGC… CD4+ T-cells SeuratPro…         70           47 0              </span></span>
<span id="cb19-31"><a href="#cb19-31"></a><span class="co">##  2 CATG… CD8+ T-cells SeuratPro…         85           52 0              </span></span>
<span id="cb19-32"><a href="#cb19-32"></a><span class="co">##  3 GAAC… CD8+ T-cells SeuratPro…         87           50 1              </span></span>
<span id="cb19-33"><a href="#cb19-33"></a><span class="co">##  4 TGAC… CD4+ T-cells SeuratPro…        127           56 0              </span></span>
<span id="cb19-34"><a href="#cb19-34"></a><span class="co">##  5 AGTC… CD4+ T-cells SeuratPro…        173           53 0              </span></span>
<span id="cb19-35"><a href="#cb19-35"></a><span class="co">##  6 TCTG… CD4+ T-cells SeuratPro…         70           48 0              </span></span>
<span id="cb19-36"><a href="#cb19-36"></a><span class="co">##  7 TGGT… CD4+ T-cells SeuratPro…         64           36 0              </span></span>
<span id="cb19-37"><a href="#cb19-37"></a><span class="co">##  8 GCAG… CD4+ T-cells SeuratPro…         72           45 0              </span></span>
<span id="cb19-38"><a href="#cb19-38"></a><span class="co">##  9 GATA… CD4+ T-cells SeuratPro…         52           36 0              </span></span>
<span id="cb19-39"><a href="#cb19-39"></a><span class="co">## 10 AATG… CD4+ T-cells SeuratPro…        100           41 0              </span></span>
<span id="cb19-40"><a href="#cb19-40"></a><span class="co">## # … with 70 more rows, and 19 more variables: letter.idents &lt;fct&gt;,</span></span>
<span id="cb19-41"><a href="#cb19-41"></a><span class="co">## #   groups &lt;chr&gt;, RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;, sample &lt;chr&gt;,</span></span>
<span id="cb19-42"><a href="#cb19-42"></a><span class="co">## #   nCount_SCT &lt;dbl&gt;, nFeature_SCT &lt;int&gt;, SCT_snn_res.0.8 &lt;fct&gt;,</span></span>
<span id="cb19-43"><a href="#cb19-43"></a><span class="co">## #   seurat_clusters &lt;fct&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;,</span></span>
<span id="cb19-44"><a href="#cb19-44"></a><span class="co">## #   PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;, UMAP_1 &lt;dbl&gt;, UMAP_2 &lt;dbl&gt;,</span></span>
<span id="cb19-45"><a href="#cb19-45"></a><span class="co">## #   UMAP_3 &lt;dbl&gt;</span></span></code></pre></div>
<p>We can easily summarise the results. For example, we can see how cell type classification overlaps with cluster classification.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>pbmc_small_cell_type <span class="op">%&gt;%</span></span>
<span id="cb20-2"><a href="#cb20-2"></a><span class="st">  </span><span class="kw">count</span>(seurat_clusters, first.labels)</span>
<span id="cb20-3"><a href="#cb20-3"></a></span>
<span id="cb20-4"><a href="#cb20-4"></a><span class="co">## # A tibble: 9 x 3</span></span>
<span id="cb20-5"><a href="#cb20-5"></a><span class="co">##   seurat_clusters first.labels     n</span></span>
<span id="cb20-6"><a href="#cb20-6"></a><span class="co">##   &lt;fct&gt;           &lt;chr&gt;        &lt;int&gt;</span></span>
<span id="cb20-7"><a href="#cb20-7"></a><span class="co">## 1 0               CD4+ T-cells     8</span></span>
<span id="cb20-8"><a href="#cb20-8"></a><span class="co">## 2 0               CD8+ T-cells    10</span></span>
<span id="cb20-9"><a href="#cb20-9"></a><span class="co">## 3 0               NK cells        12</span></span>
<span id="cb20-10"><a href="#cb20-10"></a><span class="co">## 4 1               Macrophages      1</span></span>
<span id="cb20-11"><a href="#cb20-11"></a><span class="co">## 5 1               Monocytes       25</span></span>
<span id="cb20-12"><a href="#cb20-12"></a><span class="co">## 6 2               B-cells         10</span></span>
<span id="cb20-13"><a href="#cb20-13"></a><span class="co">## 7 2               Macrophages      1</span></span>
<span id="cb20-14"><a href="#cb20-14"></a><span class="co">## 8 2               Monocytes        4</span></span>
<span id="cb20-15"><a href="#cb20-15"></a><span class="co">## 9 3               Erythrocytes     9</span></span></code></pre></div>
<p>We can easily reshape the data for building information-rich faceted plots.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a>pbmc_small_cell_type <span class="op">%&gt;%</span></span>
<span id="cb21-2"><a href="#cb21-2"></a></span>
<span id="cb21-3"><a href="#cb21-3"></a><span class="st">  </span><span class="co"># Reshape and add classifier column</span></span>
<span id="cb21-4"><a href="#cb21-4"></a><span class="st">  </span><span class="kw">pivot_longer</span>(</span>
<span id="cb21-5"><a href="#cb21-5"></a>    <span class="dt">cols =</span> <span class="kw">c</span>(seurat_clusters, first.labels),</span>
<span id="cb21-6"><a href="#cb21-6"></a>    <span class="dt">names_to =</span> <span class="st">"classifier"</span>, <span class="dt">values_to =</span> <span class="st">"label"</span></span>
<span id="cb21-7"><a href="#cb21-7"></a>  ) <span class="op">%&gt;%</span></span>
<span id="cb21-8"><a href="#cb21-8"></a></span>
<span id="cb21-9"><a href="#cb21-9"></a><span class="st">  </span><span class="co"># UMAP plots for cell type and cluster</span></span>
<span id="cb21-10"><a href="#cb21-10"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(UMAP_<span class="dv">1</span>, UMAP_<span class="dv">2</span>, <span class="dt">color =</span> label)) <span class="op">+</span></span>
<span id="cb21-11"><a href="#cb21-11"></a><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></span>
<span id="cb21-12"><a href="#cb21-12"></a><span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span>classifier) <span class="op">+</span></span>
<span id="cb21-13"><a href="#cb21-13"></a><span class="st">  </span>my_theme</span></code></pre></div>
<p><img src="reference/figures/unnamed-chunk-19-1.png"><!-- --></p>
<p>We can easily plot gene correlation per cell category, adding multi-layer annotations.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a>pbmc_small_cell_type <span class="op">%&gt;%</span></span>
<span id="cb22-2"><a href="#cb22-2"></a></span>
<span id="cb22-3"><a href="#cb22-3"></a><span class="st">  </span><span class="co"># Add some mitochondrial abundance values</span></span>
<span id="cb22-4"><a href="#cb22-4"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">mitochondrial =</span> <span class="kw">rnorm</span>(<span class="kw">n</span>())) <span class="op">%&gt;%</span></span>
<span id="cb22-5"><a href="#cb22-5"></a></span>
<span id="cb22-6"><a href="#cb22-6"></a><span class="st">  </span><span class="co"># Plot correlation</span></span>
<span id="cb22-7"><a href="#cb22-7"></a><span class="st">  </span><span class="kw">join_transcripts</span>(<span class="dt">transcripts =</span> <span class="kw">c</span>(<span class="st">"CST3"</span>, <span class="st">"LYZ"</span>), <span class="dt">shape =</span> <span class="st">"wide"</span>) <span class="op">%&gt;%</span></span>
<span id="cb22-8"><a href="#cb22-8"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(CST3 <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, LYZ <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, <span class="dt">color =</span> groups, <span class="dt">size =</span> mitochondrial)) <span class="op">+</span></span>
<span id="cb22-9"><a href="#cb22-9"></a><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></span>
<span id="cb22-10"><a href="#cb22-10"></a><span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span>first.labels, <span class="dt">scales =</span> <span class="st">"free"</span>) <span class="op">+</span></span>
<span id="cb22-11"><a href="#cb22-11"></a><span class="st">  </span><span class="kw">scale_x_log10</span>() <span class="op">+</span></span>
<span id="cb22-12"><a href="#cb22-12"></a><span class="st">  </span><span class="kw">scale_y_log10</span>() <span class="op">+</span></span>
<span id="cb22-13"><a href="#cb22-13"></a><span class="st">  </span>my_theme</span></code></pre></div>
<p><img src="reference/figures/unnamed-chunk-20-1.png"><!-- --></p>
</div>
</div>
<div id="nested-analyses" class="section level1">
<h1 class="hasAnchor">
<a href="#nested-analyses" class="anchor"></a>Nested analyses</h1>
<p>A powerful tool we can use with tidyseurat is <code>nest</code>. We can easily perform independent analyses on subsets of the dataset. First we classify cell types in lymphoid and myeloid; then, nest based on the new classification</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a>pbmc_small_nested &lt;-</span>
<span id="cb23-2"><a href="#cb23-2"></a><span class="st">  </span>pbmc_small_cell_type <span class="op">%&gt;%</span></span>
<span id="cb23-3"><a href="#cb23-3"></a><span class="st">  </span><span class="kw">filter</span>(first.labels <span class="op">!=</span><span class="st"> "Erythrocytes"</span>) <span class="op">%&gt;%</span></span>
<span id="cb23-4"><a href="#cb23-4"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">cell_class =</span> <span class="kw">if_else</span>(<span class="st">`</span><span class="dt">first.labels</span><span class="st">`</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">"Macrophages"</span>, <span class="st">"Monocytes"</span>), <span class="st">"myeloid"</span>, <span class="st">"lymphoid"</span>)) <span class="op">%&gt;%</span></span>
<span id="cb23-5"><a href="#cb23-5"></a><span class="st">  </span><span class="kw">nest</span>(<span class="dt">data =</span> <span class="op">-</span>cell_class)</span>
<span id="cb23-6"><a href="#cb23-6"></a></span>
<span id="cb23-7"><a href="#cb23-7"></a>pbmc_small_nested</span>
<span id="cb23-8"><a href="#cb23-8"></a></span>
<span id="cb23-9"><a href="#cb23-9"></a><span class="co">## # A tibble: 2 x 2</span></span>
<span id="cb23-10"><a href="#cb23-10"></a><span class="co">##   cell_class data      </span></span>
<span id="cb23-11"><a href="#cb23-11"></a><span class="co">##   &lt;chr&gt;      &lt;list&gt;    </span></span>
<span id="cb23-12"><a href="#cb23-12"></a><span class="co">## 1 lymphoid   &lt;tidysert&gt;</span></span>
<span id="cb23-13"><a href="#cb23-13"></a><span class="co">## 2 myeloid    &lt;tidysert&gt;</span></span></code></pre></div>
<p>Now we can independently for the lymphoid and myeloid subsets (i) find variable features, (ii) reduce dimensions, and (iii) cluster using both tidyverse and SingleCellExperiment seamlessly.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a>pbmc_small_nested_reanalysed &lt;-</span>
<span id="cb24-2"><a href="#cb24-2"></a><span class="st">  </span>pbmc_small_nested <span class="op">%&gt;%</span></span>
<span id="cb24-3"><a href="#cb24-3"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">data =</span> <span class="kw">map</span>(</span>
<span id="cb24-4"><a href="#cb24-4"></a>    data, <span class="op">~</span><span class="st"> </span>.x <span class="op">%&gt;%</span></span>
<span id="cb24-5"><a href="#cb24-5"></a><span class="st">      </span><span class="kw">FindVariableFeatures</span>(<span class="dt">verbose =</span> <span class="ot">FALSE</span>) <span class="op">%&gt;%</span></span>
<span id="cb24-6"><a href="#cb24-6"></a><span class="st">      </span><span class="kw">RunPCA</span>(<span class="dt">npcs =</span> <span class="dv">10</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>) <span class="op">%&gt;%</span></span>
<span id="cb24-7"><a href="#cb24-7"></a><span class="st">      </span><span class="kw">FindNeighbors</span>(<span class="dt">verbose =</span> <span class="ot">FALSE</span>) <span class="op">%&gt;%</span></span>
<span id="cb24-8"><a href="#cb24-8"></a><span class="st">      </span><span class="kw">FindClusters</span>(<span class="dt">method =</span> <span class="st">"igraph"</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>) <span class="op">%&gt;%</span></span>
<span id="cb24-9"><a href="#cb24-9"></a><span class="st">      </span><span class="kw">RunUMAP</span>(<span class="dt">reduction =</span> <span class="st">"pca"</span>, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">n.components =</span> 3L, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span>
<span id="cb24-10"><a href="#cb24-10"></a>  ))</span>
<span id="cb24-11"><a href="#cb24-11"></a></span>
<span id="cb24-12"><a href="#cb24-12"></a>pbmc_small_nested_reanalysed</span>
<span id="cb24-13"><a href="#cb24-13"></a></span>
<span id="cb24-14"><a href="#cb24-14"></a><span class="co">## # A tibble: 2 x 2</span></span>
<span id="cb24-15"><a href="#cb24-15"></a><span class="co">##   cell_class data      </span></span>
<span id="cb24-16"><a href="#cb24-16"></a><span class="co">##   &lt;chr&gt;      &lt;list&gt;    </span></span>
<span id="cb24-17"><a href="#cb24-17"></a><span class="co">## 1 lymphoid   &lt;tidysert&gt;</span></span>
<span id="cb24-18"><a href="#cb24-18"></a><span class="co">## 2 myeloid    &lt;tidysert&gt;</span></span></code></pre></div>
<p>Now we can unnest and plot the new classification.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a>pbmc_small_nested_reanalysed <span class="op">%&gt;%</span></span>
<span id="cb25-2"><a href="#cb25-2"></a></span>
<span id="cb25-3"><a href="#cb25-3"></a><span class="st">  </span><span class="co"># Convert to tibble otherwise Seurat drops reduced dimensions when unifying data sets.</span></span>
<span id="cb25-4"><a href="#cb25-4"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">data =</span> <span class="kw">map</span>(data, <span class="op">~</span><span class="st"> </span>.x <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as_tibble</span>())) <span class="op">%&gt;%</span></span>
<span id="cb25-5"><a href="#cb25-5"></a><span class="st">  </span><span class="kw">unnest</span>(data) <span class="op">%&gt;%</span></span>
<span id="cb25-6"><a href="#cb25-6"></a></span>
<span id="cb25-7"><a href="#cb25-7"></a><span class="st">  </span><span class="co"># Define unique clusters</span></span>
<span id="cb25-8"><a href="#cb25-8"></a><span class="st">  </span><span class="kw">unite</span>(<span class="st">"cluster"</span>, <span class="kw">c</span>(cell_class, seurat_clusters), <span class="dt">remove =</span> <span class="ot">FALSE</span>) <span class="op">%&gt;%</span></span>
<span id="cb25-9"><a href="#cb25-9"></a></span>
<span id="cb25-10"><a href="#cb25-10"></a><span class="st">  </span><span class="co"># Plotting</span></span>
<span id="cb25-11"><a href="#cb25-11"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(UMAP_<span class="dv">1</span>, UMAP_<span class="dv">2</span>, <span class="dt">color =</span> cluster)) <span class="op">+</span></span>
<span id="cb25-12"><a href="#cb25-12"></a><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></span>
<span id="cb25-13"><a href="#cb25-13"></a><span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span>cell_class) <span class="op">+</span></span>
<span id="cb25-14"><a href="#cb25-14"></a><span class="st">  </span>my_theme</span></code></pre></div>
<p><img src="reference/figures/unnamed-chunk-23-1.png"><!-- --></p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from CRAN at <br><a href="https://cloud.r-project.org/package=tidyseurat">https://​cloud.r-project.org/​package=tidyseurat</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Stefano Mangiola <br><small class="roles"> Author, maintainer </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.tidyverse.org/lifecycle/#maturing"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a></li>
<li><a href="https://github.com/stemangiola/tidyseurat/actions"><img src="https://github.com/stemangiola/tidyseurat/workflows/R-CMD-check-bioc/badge.svg" alt="R build status"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
