<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Brings transcriptomics to the tidyverse  • tidyseurat</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Brings transcriptomics to the tidyverse ">
<meta property="og:description" content="This is a collection of utility functions that allow
    to perform exploration of and calculations to RNA sequencing data, in
    a modular, pipe-friendly and tidy fashion.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">tidyseurat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/introduction.html">Overview of the tidyseurat package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="tidyseurat---part-of-tidytranscriptomics" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#tidyseurat---part-of-tidytranscriptomics" class="anchor"></a>tidyseurat - part of tidyTranscriptomics</h1></div>
<!-- badges: start -->

<p><strong>Brings Seurat to the tidyverse!</strong></p>
<p>website: <a href="http://stemangiola.github.io/tidyseurat/">stemangiola.github.io/tidyseurat/</a></p>
<p>Please have a look also to</p>
<ul>
<li>
<a href="https://github.com/stemangiola/nanny">nanny</a> for tidy high-level data analysis and manipulation</li>
<li>
<a href="https://github.com/stemangiola/tidygate">tidygate</a> for adding custom gate information to your tibble</li>
<li>
<a href="https://github.com/stemangiola/tidyHeatmap">tidyHeatmap</a> for heatmaps produced with tidy principles</li>
</ul>
<!---

[![Build Status](https://travis-ci.org/stemangiola/tidyseurat.svg?branch=master)](https://travis-ci.org/stemangiola/tidyseurat) [![Coverage Status](https://coveralls.io/repos/github/stemangiola/tidyseurat/badge.svg?branch=master)](https://coveralls.io/github/stemangiola/tidyseurat?branch=master)

--><div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>From CRAN (in submission)</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"tidyseurat"</span>)
</pre></div>
<p>From Github</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="kw">devtools</span>::<span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"stemangiola/tidyseurat"</span>)
</pre></div>
</div>
<div id="create-tidyseurat" class="section level2">
<h2 class="hasAnchor">
<a href="#create-tidyseurat" class="anchor"></a>Create <code>tidyseurat</code>
</h2>
<p>This is a seurat object but it is evaluated as tibble. So it is fully compatible both with Seurat and tidyverse APIs.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="kw">pbmc_small_tidy</span> <span class="op">=</span> <span class="kw">tidyseurat</span>::<span class="kw"><a href="reference/pbmc_small.html">pbmc_small</a></span> <span class="op">%&gt;%</span> <span class="fu"><a href="reference/tidy.html">tidy</a></span>()
<span class="kw">pbmc_small_tidy</span>
</pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="co">## # A tibble: 80 x 15</span></span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="co">##    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups</span></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="co">##    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; </span></span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="co">##  1 ATGC… SeuratPro…         70           47 0               A             g2    </span></span>
<span id="cb4-5"><a href="#cb4-5"></a><span class="co">##  2 CATG… SeuratPro…         85           52 0               A             g1    </span></span>
<span id="cb4-6"><a href="#cb4-6"></a><span class="co">##  3 GAAC… SeuratPro…         87           50 1               B             g2    </span></span>
<span id="cb4-7"><a href="#cb4-7"></a><span class="co">##  4 TGAC… SeuratPro…        127           56 0               A             g2    </span></span>
<span id="cb4-8"><a href="#cb4-8"></a><span class="co">##  5 AGTC… SeuratPro…        173           53 0               A             g2    </span></span>
<span id="cb4-9"><a href="#cb4-9"></a><span class="co">##  6 TCTG… SeuratPro…         70           48 0               A             g1    </span></span>
<span id="cb4-10"><a href="#cb4-10"></a><span class="co">##  7 TGGT… SeuratPro…         64           36 0               A             g1    </span></span>
<span id="cb4-11"><a href="#cb4-11"></a><span class="co">##  8 GCAG… SeuratPro…         72           45 0               A             g1    </span></span>
<span id="cb4-12"><a href="#cb4-12"></a><span class="co">##  9 GATA… SeuratPro…         52           36 0               A             g1    </span></span>
<span id="cb4-13"><a href="#cb4-13"></a><span class="co">## 10 AATG… SeuratPro…        100           41 0               A             g1    </span></span>
<span id="cb4-14"><a href="#cb4-14"></a><span class="co">## # … with 70 more rows, and 8 more variables: RNA_snn_res.1 &lt;fct&gt;, PC_1 &lt;dbl&gt;,</span></span>
<span id="cb4-15"><a href="#cb4-15"></a><span class="co">## #   PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</span></span></code></pre></div>
</div>
<div id="preliminary-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#preliminary-plots" class="anchor"></a>Preliminary plots</h2>
<p>We can treat <code>pbmc_small_tidy</code> effectively as a normal tibble.</p>
<p>Here we plot number of transcripts per cell</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="kw">pbmc_small_tidy</span> <span class="op">%&gt;%</span>
  <span class="kw">tidyseurat</span>::<span class="fu"><a href="reference/ggplot.html">ggplot</a></span>(<span class="fu">aes</span>(<span class="kw">nFeature_RNA</span>, fill=<span class="kw">groups</span>)) <span class="op">+</span> 
  <span class="fu">geom_histogram</span>() <span class="op">+</span>
  <span class="kw">my_theme</span>
</pre></div>
<p><img src="reference/figures/plot1-1.png"><!-- --></p>
<p>Here we plot total transcriptional material per cell</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="kw">pbmc_small_tidy</span> <span class="op">%&gt;%</span>
  <span class="kw">tidyseurat</span>::<span class="fu"><a href="reference/ggplot.html">ggplot</a></span>(<span class="fu">aes</span>(<span class="kw">groups</span>, <span class="kw">nCount_RNA</span>, fill=<span class="kw">groups</span>)) <span class="op">+</span> 
  <span class="fu">geom_boxplot</span>(outlier.shape = <span class="fl">NA</span>) <span class="op">+</span>
  <span class="fu">geom_jitter</span>(width = <span class="fl">0.1</span>) <span class="op">+</span>
  <span class="kw">my_theme</span>
</pre></div>
<p><img src="reference/figures/plot2-1.png"><!-- --></p>
<p>Here we plot abundance of two transcripts for each group</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="kw">pbmc_small_tidy</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/join_transcripts.html">join_transcripts</a></span>(transcripts = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"HLA-DRA"</span> ,     <span class="st">"LYZ"</span> )) <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/ggplot.html">ggplot</a></span>(<span class="fu">aes</span>(<span class="kw">groups</span>, <span class="kw">abundance_RNA</span> <span class="op">+</span> <span class="fl">1</span>, fill=<span class="kw">groups</span>)) <span class="op">+</span> 
  <span class="fu">geom_boxplot</span>(outlier.shape = <span class="fl">NA</span>) <span class="op">+</span> 
  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(size=<span class="kw">nCount_RNA</span>), alpha=<span class="fl">0.5</span>, width = <span class="fl">0.2</span>) <span class="op">+</span> 
  <span class="fu">scale_y_log10</span>() <span class="op">+</span> 
  <span class="kw">my_theme</span>
</pre></div>
<p><img src="reference/figures/unnamed-chunk-6-1.png"><!-- --></p>
</div>
<div id="preprocess-the-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#preprocess-the-dataset" class="anchor"></a>Preprocess the dataset</h2>
<p>Also you can treat the object as Seurat object and proceed with data processing.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="kw">pbmc_small_pca</span> <span class="op">=</span> 
  <span class="kw">pbmc_small_tidy</span> <span class="op">%&gt;%</span>
  <span class="fu">SCTransform</span>(verbose = <span class="fl">FALSE</span>) <span class="op">%&gt;%</span> 
  <span class="fu">FindVariableFeatures</span>(verbose = <span class="fl">FALSE</span>) <span class="op">%&gt;%</span> 
  <span class="fu">RunPCA</span>(verbose = <span class="fl">FALSE</span>)

<span class="kw">pbmc_small_pca</span>
</pre></div>
<div class="sourceCode" id="cb9"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="co">## # A tibble: 80 x 17</span></span>
<span id="cb9-2"><a href="#cb9-2"></a><span class="co">##    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups</span></span>
<span id="cb9-3"><a href="#cb9-3"></a><span class="co">##    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; </span></span>
<span id="cb9-4"><a href="#cb9-4"></a><span class="co">##  1 ATGC… SeuratPro…         70           47 0               A             g2    </span></span>
<span id="cb9-5"><a href="#cb9-5"></a><span class="co">##  2 CATG… SeuratPro…         85           52 0               A             g1    </span></span>
<span id="cb9-6"><a href="#cb9-6"></a><span class="co">##  3 GAAC… SeuratPro…         87           50 1               B             g2    </span></span>
<span id="cb9-7"><a href="#cb9-7"></a><span class="co">##  4 TGAC… SeuratPro…        127           56 0               A             g2    </span></span>
<span id="cb9-8"><a href="#cb9-8"></a><span class="co">##  5 AGTC… SeuratPro…        173           53 0               A             g2    </span></span>
<span id="cb9-9"><a href="#cb9-9"></a><span class="co">##  6 TCTG… SeuratPro…         70           48 0               A             g1    </span></span>
<span id="cb9-10"><a href="#cb9-10"></a><span class="co">##  7 TGGT… SeuratPro…         64           36 0               A             g1    </span></span>
<span id="cb9-11"><a href="#cb9-11"></a><span class="co">##  8 GCAG… SeuratPro…         72           45 0               A             g1    </span></span>
<span id="cb9-12"><a href="#cb9-12"></a><span class="co">##  9 GATA… SeuratPro…         52           36 0               A             g1    </span></span>
<span id="cb9-13"><a href="#cb9-13"></a><span class="co">## 10 AATG… SeuratPro…        100           41 0               A             g1    </span></span>
<span id="cb9-14"><a href="#cb9-14"></a><span class="co">## # … with 70 more rows, and 10 more variables: RNA_snn_res.1 &lt;fct&gt;,</span></span>
<span id="cb9-15"><a href="#cb9-15"></a><span class="co">## #   nCount_SCT &lt;dbl&gt;, nFeature_SCT &lt;int&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;,</span></span>
<span id="cb9-16"><a href="#cb9-16"></a><span class="co">## #   PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</span></span></code></pre></div>
<p>If a tool is not included in the tidyseurat collection, we can use <code>as_tibble</code> to permanently convert <code>tidyseurat</code> into tibble</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="kw">pbmc_small_pca</span> <span class="op">%&gt;%</span>
  <span class="kw">as_tibble</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/select.html">select</a></span>(<span class="fu">contains</span>(<span class="st">"PC"</span>), <span class="fu">everything</span>()) <span class="op">%&gt;%</span>
      <span class="kw">GGally</span>::<span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html">ggpairs</a></span>(columns = <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="kw">ggplot2</span>::<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(colour=<span class="kw">groups</span>))
</pre></div>
<p><img src="reference/figures/pc%20plot-1.png"><!-- --></p>
</div>
<div id="identify-clusters-and-reduce-dimensions" class="section level2">
<h2 class="hasAnchor">
<a href="#identify-clusters-and-reduce-dimensions" class="anchor"></a>Identify clusters and reduce dimensions</h2>
<p>We proceed with cluster identification</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="kw">pbmc_small_cluster</span> <span class="op">=</span> 
  <span class="kw">pbmc_small_pca</span> <span class="op">%&gt;%</span>
  <span class="fu">FindNeighbors</span>(verbose=<span class="fl">FALSE</span>) <span class="op">%&gt;%</span> 
  <span class="fu">FindClusters</span>(method = <span class="st">"igraph"</span>, verbose=<span class="fl">FALSE</span> )

<span class="kw">pbmc_small_cluster</span>
</pre></div>
<div class="sourceCode" id="cb12"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="co">## # A tibble: 80 x 19</span></span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="co">##    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups</span></span>
<span id="cb12-3"><a href="#cb12-3"></a><span class="co">##    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; </span></span>
<span id="cb12-4"><a href="#cb12-4"></a><span class="co">##  1 ATGC… SeuratPro…         70           47 0               A             g2    </span></span>
<span id="cb12-5"><a href="#cb12-5"></a><span class="co">##  2 CATG… SeuratPro…         85           52 0               A             g1    </span></span>
<span id="cb12-6"><a href="#cb12-6"></a><span class="co">##  3 GAAC… SeuratPro…         87           50 1               B             g2    </span></span>
<span id="cb12-7"><a href="#cb12-7"></a><span class="co">##  4 TGAC… SeuratPro…        127           56 0               A             g2    </span></span>
<span id="cb12-8"><a href="#cb12-8"></a><span class="co">##  5 AGTC… SeuratPro…        173           53 0               A             g2    </span></span>
<span id="cb12-9"><a href="#cb12-9"></a><span class="co">##  6 TCTG… SeuratPro…         70           48 0               A             g1    </span></span>
<span id="cb12-10"><a href="#cb12-10"></a><span class="co">##  7 TGGT… SeuratPro…         64           36 0               A             g1    </span></span>
<span id="cb12-11"><a href="#cb12-11"></a><span class="co">##  8 GCAG… SeuratPro…         72           45 0               A             g1    </span></span>
<span id="cb12-12"><a href="#cb12-12"></a><span class="co">##  9 GATA… SeuratPro…         52           36 0               A             g1    </span></span>
<span id="cb12-13"><a href="#cb12-13"></a><span class="co">## 10 AATG… SeuratPro…        100           41 0               A             g1    </span></span>
<span id="cb12-14"><a href="#cb12-14"></a><span class="co">## # … with 70 more rows, and 12 more variables: RNA_snn_res.1 &lt;fct&gt;,</span></span>
<span id="cb12-15"><a href="#cb12-15"></a><span class="co">## #   nCount_SCT &lt;dbl&gt;, nFeature_SCT &lt;int&gt;, SCT_snn_res.0.8 &lt;fct&gt;,</span></span>
<span id="cb12-16"><a href="#cb12-16"></a><span class="co">## #   seurat_clusters &lt;fct&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;,</span></span>
<span id="cb12-17"><a href="#cb12-17"></a><span class="co">## #   PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</span></span></code></pre></div>
<p>Now we can interrogate the object as if it was a regular tibble data frame</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="kw">pbmc_small_cluster</span> <span class="op">%&gt;%</span>
  <span class="kw">tidyseurat</span>::<span class="fu"><a href="reference/count.html">count</a></span>(<span class="kw">groups</span>, <span class="kw">seurat_clusters</span>)
</pre></div>
<div class="sourceCode" id="cb14"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="co">## # A tibble: 8 x 3</span></span>
<span id="cb14-2"><a href="#cb14-2"></a><span class="co">## # Groups:   groups [2]</span></span>
<span id="cb14-3"><a href="#cb14-3"></a><span class="co">##   groups seurat_clusters     n</span></span>
<span id="cb14-4"><a href="#cb14-4"></a><span class="co">##   &lt;chr&gt;  &lt;fct&gt;           &lt;int&gt;</span></span>
<span id="cb14-5"><a href="#cb14-5"></a><span class="co">## 1 g1     0                  17</span></span>
<span id="cb14-6"><a href="#cb14-6"></a><span class="co">## 2 g1     1                  14</span></span>
<span id="cb14-7"><a href="#cb14-7"></a><span class="co">## 3 g1     2                   9</span></span>
<span id="cb14-8"><a href="#cb14-8"></a><span class="co">## 4 g1     3                   4</span></span>
<span id="cb14-9"><a href="#cb14-9"></a><span class="co">## 5 g2     0                  13</span></span>
<span id="cb14-10"><a href="#cb14-10"></a><span class="co">## 6 g2     1                  12</span></span>
<span id="cb14-11"><a href="#cb14-11"></a><span class="co">## 7 g2     2                   6</span></span>
<span id="cb14-12"><a href="#cb14-12"></a><span class="co">## 8 g2     3                   5</span></span></code></pre></div>
</div>
<div id="reduce-dimensions" class="section level2">
<h2 class="hasAnchor">
<a href="#reduce-dimensions" class="anchor"></a>Reduce dimensions</h2>
<p>We can calculate the first 3 UMAP dimensions using Seurat framework</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="kw">pbmc_small_UMAP</span> <span class="op">=</span> 
  <span class="kw">pbmc_small_cluster</span> <span class="op">%&gt;%</span>
  <span class="fu">RunUMAP</span>(reduction = <span class="st">"pca"</span>, dims = <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>, n.components = <span class="fl">3L</span>, )
</pre></div>
<p>and we can plot them using 3D plot using plotly</p>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="kw">pbmc_small_UMAP</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/plot_ly.html">plot_ly</a></span>(
        x = <span class="op">~</span><span class="kw">`UMAP_1`</span>,
        y = <span class="op">~</span><span class="kw">`UMAP_2`</span>, 
        z = <span class="op">~</span><span class="kw">`UMAP_3`</span>,
        color = <span class="op">~</span> <span class="kw">seurat_clusters</span>
    ) 
</pre></div>
<!--html_preserve-->
<div id="htmlwidget-c64457a584143ef1e822" class="plotly html-widget" style="width:672px;height:480px;">
</div>
<script type="application/json" data-for="htmlwidget-c64457a584143ef1e822">{"x":{"visdat":{"133d25efe0b6":["function () ","plotlyVisDat"]},"cur_data":"133d25efe0b6","attrs":{"133d25efe0b6":{"x":{},"y":{},"z":{},"color":{},"alpha_stroke":1,"sizes":[10,100],"spans":[1,20]}},"layout":{"margin":{"b":40,"l":60,"t":25,"r":10},"scene":{"xaxis":{"title":"UMAP_1"},"yaxis":{"title":"UMAP_2"},"zaxis":{"title":"UMAP_3"}},"hovermode":"closest","showlegend":true},"source":"A","config":{"showSendToCloud":false},"data":[{"x":[6.53173612952232,6.36876940131187,7.03650164008141,6.55522131323814,6.6794950902462,7.01866983771324,6.7821824491024,6.40633463263512,6.59762024283409,6.18055986762047,6.9664938390255,7.01296066641808,6.49511312842369,6.56305193305016,6.57743000388145,6.89375137686729,7.01308559775353,6.48464416861534,7.10751509070396,6.93657039999962,6.92854380011559,7.28502630591393,6.71771311163902,7.29223131537437,7.04746269583702,6.7464272916317,7.30604767203331,7.25595402121544,7.01022886633873,7.06775307059288],"y":[11.5283135354519,10.7085894525051,10.8671339929104,11.257818120718,10.5868786752224,11.038299459219,11.1493634164333,11.2481860101223,11.1395472466946,11.0592984139919,7.59875430464745,8.18448104262352,8.23392619490624,8.08994473814964,7.74761618971825,8.01899804472923,7.64437903761864,7.60362996459007,8.14712991118431,7.81488646864891,8.6157511651516,9.41175355315209,9.94269217848778,9.44877995848656,9.99967660307884,9.46758260130882,8.68599786162376,9.80573644042015,9.03281535506249,10.0702938020229],"z":[-5.56003659069538,-5.13579886257648,-5.25702565014362,-4.94951527416706,-5.18163292706013,-5.73782055675983,-6.02730362713337,-5.72730629742146,-5.29015534222126,-5.49146549999714,-6.50661270916462,-5.87316363155842,-6.49952547848225,-6.20535891354084,-6.28932850658894,-6.93289082348347,-6.86140053570271,-6.82902186214924,-6.71219961941242,-6.14218752682209,-6.22294466793537,-6.19121640026569,-5.78895991146564,-5.77883475124836,-5.54875557720661,-5.84484141170979,-5.94413988888264,-6.16199057400227,-5.94984810650349,-6.03364651501179],"type":"scatter3d","mode":"markers","name":"0","marker":{"color":"rgba(102,194,165,1)","line":{"color":"rgba(102,194,165,1)"}},"textfont":{"color":"rgba(102,194,165,1)"},"error_y":{"color":"rgba(102,194,165,1)"},"error_x":{"color":"rgba(102,194,165,1)"},"line":{"color":"rgba(102,194,165,1)"},"frame":null},{"x":[-9.57227254509926,-8.64794278740883,-9.61629796624184,-9.32994485497475,-9.60987020134926,-9.38935018181801,-9.16148973107338,-9.64048219323158,-8.96156621575355,-9.07710958123207,-9.40074754357338,-10.1616547167301,-9.37781739830971,-9.87877202630043,-9.98006177544594,-9.34374356865883,-9.97350335717201,-9.64729905724525,-9.69794488549232,-9.74219346642494,-8.47162461876869,-8.2341110765934,-7.86443448662758,-8.68385053277016,-8.44186139702797,-8.6944267809391],"y":[-6.23931418061256,-6.56876764893532,-7.08914480805397,-6.78720770478249,-6.66331587433815,-6.84464750885963,-6.41740999817848,-6.64174757599831,-6.97174654603004,-7.16103945374489,-6.80577765107155,-7.28446112275124,-7.09383974671364,-7.1702185690403,-6.76543055176735,-6.90914641022682,-7.07067404389381,-7.08411131501198,-6.80779562592506,-7.48406992554665,-5.09724579453468,-5.74595747590065,-5.51403103470802,-5.97120485901833,-5.50579605698586,-6.64153872132301],"z":[3.00410539805889,3.67709214389324,2.91037852466106,2.75606543719769,3.14085704982281,3.37742836177349,3.24494416415691,2.98054630458355,3.26856596171856,3.56985146701336,4.15854317843914,4.8109775274992,4.62381655871868,4.52742821872234,5.033973762393,4.91579062640667,5.14831406772137,5.30594832599163,4.91577441394329,4.90894705951214,2.82102329432964,2.96294004619122,2.52153880298138,2.80141813457012,2.47882635295391,2.7987590521574],"type":"scatter3d","mode":"markers","name":"1","marker":{"color":"rgba(252,141,98,1)","line":{"color":"rgba(252,141,98,1)"}},"textfont":{"color":"rgba(252,141,98,1)"},"error_y":{"color":"rgba(252,141,98,1)"},"error_x":{"color":"rgba(252,141,98,1)"},"line":{"color":"rgba(252,141,98,1)"},"frame":null},{"x":[-6.88667131066322,-7.07096791863441,-7.4671900331974,-7.19372678399086,-7.38110471367836,-7.6712276995182,-7.32879186272621,-7.07979798913002,-7.60068441033363,-7.40269399285316,-7.61055493950844,-7.86260057091713,-8.27942681908607,-8.05179906487465,-8.10963178277016],"y":[-3.00643311142921,-2.41199598908424,-3.10626183152199,-2.80095348954201,-2.58702717423439,-2.58593903183937,-3.19867096543312,-2.70369158387184,-2.94522057175636,-2.60775528550148,-5.18772946000099,-5.30893097519875,-4.58480797410011,-5.12688456177712,-5.18986807465553],"z":[3.94610411822796,3.61617333590984,3.70458132922649,4.24318558871746,3.37905080020428,3.69899518191814,4.09670693576336,3.78391987979412,4.18364436328411,4.10225732028484,2.61790830790997,2.92269904315472,2.77718932330608,3.00402028262615,2.45375282466412],"type":"scatter3d","mode":"markers","name":"2","marker":{"color":"rgba(141,160,203,1)","line":{"color":"rgba(141,160,203,1)"}},"textfont":{"color":"rgba(141,160,203,1)"},"error_y":{"color":"rgba(141,160,203,1)"},"error_x":{"color":"rgba(141,160,203,1)"},"line":{"color":"rgba(141,160,203,1)"},"frame":null},{"x":[16.7370288312435,16.0774247586727,16.742524856329,16.9703032910824,16.3220775067806,16.661452049017,16.4785411298275,16.5126397550106,16.2301557004452],"y":[-6.75542078614235,-6.49676142334938,-6.52306089997292,-6.22149191498756,-6.76576338410378,-6.36998568177223,-6.39005480408669,-6.60305796265602,-6.33484659790993],"z":[2.97382862269878,3.14184601008892,3.55932576358318,3.30012113749981,3.41675026118755,2.94473392665386,3.55637175738811,3.21299035251141,3.37491733729839],"type":"scatter3d","mode":"markers","name":"3","marker":{"color":"rgba(231,138,195,1)","line":{"color":"rgba(231,138,195,1)"}},"textfont":{"color":"rgba(231,138,195,1)"},"error_y":{"color":"rgba(231,138,195,1)"},"error_x":{"color":"rgba(231,138,195,1)"},"line":{"color":"rgba(231,138,195,1)"},"frame":null}],"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script><!--/html_preserve-->
</div>
<div id="cell-type-prediction" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-type-prediction" class="anchor"></a>Cell type prediction</h2>
<p>We can infer cell type identities using singleR</p>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="kw">blueprint</span> <span class="op">=</span> <span class="kw">SingleR</span>::<span class="fu">BlueprintEncodeData</span>()

<span class="kw">cell_type_df</span> <span class="op">=</span> 
  <span class="kw">pbmc_small_UMAP</span><span class="op">@</span>assays[[<span class="st">"SCT"</span>]]<span class="op">@</span>counts <span class="op">%&gt;%</span> 
  <span class="kw">log1p</span> <span class="op">%&gt;%</span>
  <span class="kw">Matrix</span>::<span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/Matrix.html">Matrix</a></span>(sparse = <span class="fl">TRUE</span>) <span class="op">%&gt;%</span>
 <span class="kw">SingleR</span>::<span class="fu">SingleR</span>(
     ref = <span class="kw">blueprint</span> ,
     labels = <span class="kw">blueprint</span><span class="op">$</span><span class="kw">label.main</span>,
     method = <span class="st">"single"</span>
 ) <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>() <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/as_tibble.html">as_tibble</a></span>(rownames=<span class="st">"cell"</span>) <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/select.html">select</a></span>(<span class="kw">cell</span>, <span class="kw">first.labels</span>)
</pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit">
<span class="kw">pbmc_small_cell_type</span> <span class="op">=</span>
  <span class="kw">pbmc_small_UMAP</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/left_join.html">left_join</a></span>(<span class="kw">cell_type_df</span>, by=<span class="st">"cell"</span>)

<span class="kw">pbmc_small_cell_type</span> <span class="op">%&gt;%</span>
  <span class="kw">tidyseurat</span>::<span class="fu"><a href="reference/select.html">select</a></span>(<span class="kw">cell</span>, <span class="kw">first.labels</span>, <span class="fu">everything</span>())
</pre></div>
<div class="sourceCode" id="cb19"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a><span class="co">## # A tibble: 80 x 23</span></span>
<span id="cb19-2"><a href="#cb19-2"></a><span class="co">##    cell  first.labels orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8</span></span>
<span id="cb19-3"><a href="#cb19-3"></a><span class="co">##    &lt;chr&gt; &lt;chr&gt;        &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;          </span></span>
<span id="cb19-4"><a href="#cb19-4"></a><span class="co">##  1 ATGC… CD4+ T-cells SeuratPro…         70           47 0              </span></span>
<span id="cb19-5"><a href="#cb19-5"></a><span class="co">##  2 CATG… CD8+ T-cells SeuratPro…         85           52 0              </span></span>
<span id="cb19-6"><a href="#cb19-6"></a><span class="co">##  3 GAAC… CD8+ T-cells SeuratPro…         87           50 1              </span></span>
<span id="cb19-7"><a href="#cb19-7"></a><span class="co">##  4 TGAC… CD4+ T-cells SeuratPro…        127           56 0              </span></span>
<span id="cb19-8"><a href="#cb19-8"></a><span class="co">##  5 AGTC… CD4+ T-cells SeuratPro…        173           53 0              </span></span>
<span id="cb19-9"><a href="#cb19-9"></a><span class="co">##  6 TCTG… CD4+ T-cells SeuratPro…         70           48 0              </span></span>
<span id="cb19-10"><a href="#cb19-10"></a><span class="co">##  7 TGGT… CD4+ T-cells SeuratPro…         64           36 0              </span></span>
<span id="cb19-11"><a href="#cb19-11"></a><span class="co">##  8 GCAG… CD4+ T-cells SeuratPro…         72           45 0              </span></span>
<span id="cb19-12"><a href="#cb19-12"></a><span class="co">##  9 GATA… CD4+ T-cells SeuratPro…         52           36 0              </span></span>
<span id="cb19-13"><a href="#cb19-13"></a><span class="co">## 10 AATG… CD4+ T-cells SeuratPro…        100           41 0              </span></span>
<span id="cb19-14"><a href="#cb19-14"></a><span class="co">## # … with 70 more rows, and 17 more variables: letter.idents &lt;fct&gt;,</span></span>
<span id="cb19-15"><a href="#cb19-15"></a><span class="co">## #   groups &lt;chr&gt;, RNA_snn_res.1 &lt;fct&gt;, nCount_SCT &lt;dbl&gt;, nFeature_SCT &lt;int&gt;,</span></span>
<span id="cb19-16"><a href="#cb19-16"></a><span class="co">## #   SCT_snn_res.0.8 &lt;fct&gt;, seurat_clusters &lt;fct&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;,</span></span>
<span id="cb19-17"><a href="#cb19-17"></a><span class="co">## #   PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;,</span></span>
<span id="cb19-18"><a href="#cb19-18"></a><span class="co">## #   UMAP_1 &lt;dbl&gt;, UMAP_2 &lt;dbl&gt;, UMAP_3 &lt;dbl&gt;</span></span></code></pre></div>
<p>We can easily summarise the results. For example, see how cell type classification overlaps with cluster classification.</p>
<div class="sourceCode" id="cb20"><pre class="downlit">
<span class="kw">pbmc_small_cell_type</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/count.html">count</a></span>(<span class="kw">seurat_clusters</span>, <span class="kw">first.labels</span>)
</pre></div>
<div class="sourceCode" id="cb21"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="co">## # A tibble: 9 x 3</span></span>
<span id="cb21-2"><a href="#cb21-2"></a><span class="co">## # Groups:   seurat_clusters [4]</span></span>
<span id="cb21-3"><a href="#cb21-3"></a><span class="co">##   seurat_clusters first.labels     n</span></span>
<span id="cb21-4"><a href="#cb21-4"></a><span class="co">##   &lt;fct&gt;           &lt;chr&gt;        &lt;int&gt;</span></span>
<span id="cb21-5"><a href="#cb21-5"></a><span class="co">## 1 0               CD4+ T-cells     8</span></span>
<span id="cb21-6"><a href="#cb21-6"></a><span class="co">## 2 0               CD8+ T-cells    10</span></span>
<span id="cb21-7"><a href="#cb21-7"></a><span class="co">## 3 0               NK cells        12</span></span>
<span id="cb21-8"><a href="#cb21-8"></a><span class="co">## 4 1               Macrophages      1</span></span>
<span id="cb21-9"><a href="#cb21-9"></a><span class="co">## 5 1               Monocytes       25</span></span>
<span id="cb21-10"><a href="#cb21-10"></a><span class="co">## 6 2               B-cells         10</span></span>
<span id="cb21-11"><a href="#cb21-11"></a><span class="co">## 7 2               Macrophages      1</span></span>
<span id="cb21-12"><a href="#cb21-12"></a><span class="co">## 8 2               Monocytes        4</span></span>
<span id="cb21-13"><a href="#cb21-13"></a><span class="co">## 9 3               Erythrocytes     9</span></span></code></pre></div>
<p>We can easily reshape the data for building information-rish faceted plots</p>
<div class="sourceCode" id="cb22"><pre class="downlit">
<span class="kw">pbmc_small_cell_type</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Reshaping</span>
  <span class="fu"><a href="reference/pivot_longer.html">pivot_longer</a></span>(
    cols=<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw">seurat_clusters</span>, <span class="kw">first.labels</span>), 
    names_to = <span class="st">"classifier"</span>, values_to = <span class="st">"label"</span>
  ) <span class="op">%&gt;%</span>
  
  <span class="co"># Plotting</span>
  <span class="fu"><a href="reference/ggplot.html">ggplot</a></span>(<span class="fu">aes</span>(<span class="kw">UMAP_1</span>, <span class="kw">UMAP_2</span>, color=<span class="kw">label</span>)) <span class="op">+</span>
  <span class="fu">geom_point</span>() <span class="op">+</span>
  <span class="fu">facet_wrap</span>(<span class="op">~</span><span class="kw">classifier</span>) <span class="op">+</span>
  <span class="kw">my_theme</span>
</pre></div>
<p><img src="reference/figures/unnamed-chunk-10-1.png"><!-- --></p>
<p>We can easily plot gene correlation per cell category, adding multi-layer annotations</p>
<div class="sourceCode" id="cb23"><pre class="downlit">
<span class="kw">pbmc_small_cell_type</span> <span class="op">%&gt;%</span> 
  
  <span class="co"># Add mitochondrial abundance</span>
  <span class="fu"><a href="reference/mutate.html">mutate</a></span>(mitochondrial = <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(<span class="fu">n</span>())) <span class="op">%&gt;%</span>
  
  <span class="co"># Plot correlation</span>
  <span class="fu"><a href="reference/join_transcripts.html">join_transcripts</a></span>(transcripts = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CST3"</span> ,     <span class="st">"LYZ"</span> ), shape = <span class="st">"wide"</span>) <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/ggplot.html">ggplot</a></span>(<span class="fu">aes</span>(<span class="kw">CST3</span> <span class="op">+</span><span class="fl">1</span>, <span class="kw">LYZ</span> <span class="op">+</span> <span class="fl">1</span>, color=<span class="kw">groups</span>, size=<span class="kw">mitochondrial</span>)) <span class="op">+</span>
  <span class="fu">geom_point</span>() <span class="op">+</span> 
  <span class="fu">facet_wrap</span>(<span class="op">~</span><span class="kw">first.labels</span>, scales = <span class="st">"free"</span>) <span class="op">+</span>
  <span class="fu">scale_x_log10</span>() <span class="op">+</span>
  <span class="fu">scale_y_log10</span>() <span class="op">+</span>
  <span class="kw">my_theme</span>
</pre></div>
<p><img src="reference/figures/unnamed-chunk-11-1.png"><!-- --> ## Nested analyses A powerful tool we can use with tidyseurat is <code>nest</code>. We can easily perform independent analyses on subsets of the dataset. First we classify cell types in lymphoid and myeloid; then, nest based on the new classification</p>
<div class="sourceCode" id="cb24"><pre class="downlit">
<span class="kw">pbmc_small_nested</span> <span class="op">=</span> 
  <span class="kw">pbmc_small_cell_type</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/filter.html">filter</a></span>(<span class="kw">first.labels</span> != <span class="st">"Erythrocytes"</span>) <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/mutate.html">mutate</a></span>(cell_class = <span class="fu">if_else</span>(<span class="kw">`first.labels`</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Macrophages"</span>, <span class="st">"Monocytes"</span>), <span class="st">"myeloid"</span>, <span class="st">"lmphoid"</span>)) <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/unnest.html">nest</a></span>(data = <span class="op">-</span><span class="kw">cell_class</span>)

<span class="kw">pbmc_small_nested</span>
</pre></div>
<div class="sourceCode" id="cb25"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a><span class="co">## # A tibble: 2 x 2</span></span>
<span id="cb25-2"><a href="#cb25-2"></a><span class="co">##   cell_class data      </span></span>
<span id="cb25-3"><a href="#cb25-3"></a><span class="co">##   &lt;chr&gt;      &lt;list&gt;    </span></span>
<span id="cb25-4"><a href="#cb25-4"></a><span class="co">## 1 lmphoid    &lt;tidysert&gt;</span></span>
<span id="cb25-5"><a href="#cb25-5"></a><span class="co">## 2 myeloid    &lt;tidysert&gt;</span></span></code></pre></div>
<p>Now we can independently (i) find variable features, (ii) reduce dimensions, and (iii) cluster. Using both tidyverse and Seurat seamlessy</p>
<div class="sourceCode" id="cb26"><pre class="downlit">
<span class="kw">pbmc_small_nested_reanalysed</span> <span class="op">=</span> 
  <span class="kw">pbmc_small_nested</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/mutate.html">mutate</a></span>(data = <span class="fu">map</span>(
    <span class="kw">data</span>, <span class="op">~</span> <span class="kw">.x</span> <span class="op">%&gt;%</span> 
      <span class="fu">FindVariableFeatures</span>(verbose = <span class="fl">FALSE</span>) <span class="op">%&gt;%</span>
      <span class="fu">RunPCA</span>(npcs = <span class="fl">10</span>, verbose = <span class="fl">FALSE</span>) <span class="op">%&gt;%</span>
      <span class="fu">FindNeighbors</span>(verbose = <span class="fl">FALSE</span>) <span class="op">%&gt;%</span> 
      <span class="fu">FindClusters</span>(method = <span class="st">"igraph"</span> , verbose = <span class="fl">FALSE</span>) <span class="op">%&gt;%</span> 
      <span class="fu">RunUMAP</span>(reduction = <span class="st">"pca"</span>, dims = <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, n.components = <span class="fl">3L</span>, verbose = <span class="fl">FALSE</span>)
  )) 

<span class="kw">pbmc_small_nested_reanalysed</span>
</pre></div>
<div class="sourceCode" id="cb27"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a><span class="co">## # A tibble: 2 x 2</span></span>
<span id="cb27-2"><a href="#cb27-2"></a><span class="co">##   cell_class data      </span></span>
<span id="cb27-3"><a href="#cb27-3"></a><span class="co">##   &lt;chr&gt;      &lt;list&gt;    </span></span>
<span id="cb27-4"><a href="#cb27-4"></a><span class="co">## 1 lmphoid    &lt;tidysert&gt;</span></span>
<span id="cb27-5"><a href="#cb27-5"></a><span class="co">## 2 myeloid    &lt;tidysert&gt;</span></span></code></pre></div>
<p>Now we can unnest and plot the new clasification</p>
<div class="sourceCode" id="cb28"><pre class="downlit">
<span class="kw">pbmc_small_nested_reanalysed</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Convert to tibble otherwise Seurat drops reduced dimensions when unifying data sets.</span>
  <span class="fu"><a href="reference/mutate.html">mutate</a></span>(data = <span class="fu">map</span>(<span class="kw">data</span>, <span class="op">~</span> <span class="kw">.x</span> <span class="op">%&gt;%</span> <span class="kw">as_tibble</span>)) <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/unnest.html">unnest</a></span>(<span class="kw">data</span>) <span class="op">%&gt;%</span>

  <span class="co"># Define unique clusters</span>
  <span class="fu"><a href="reference/unite.html">unite</a></span>(<span class="st">"cluster"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw">cell_class</span>, <span class="kw">seurat_clusters</span>), remove=<span class="fl">FALSE</span>) <span class="op">%&gt;%</span>
  
  <span class="co"># Plotting</span>
  <span class="fu"><a href="reference/ggplot.html">ggplot</a></span>(<span class="fu">aes</span>(<span class="kw">UMAP_1</span>, <span class="kw">UMAP_2</span>, color=<span class="kw">cluster</span>)) <span class="op">+</span>
  <span class="fu">geom_point</span>() <span class="op">+</span>
  <span class="fu">facet_wrap</span>(<span class="op">~</span><span class="kw">cell_class</span>) <span class="op">+</span>
  <span class="kw">my_theme</span>
</pre></div>
<p><img src="reference/figures/unnamed-chunk-14-1.png"><!-- --></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Stefano Mangiola <br><small class="roles"> Author, maintainer </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.tidyverse.org/lifecycle/#maturing"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a></li>
<li><a href="https://github.com/stemangiola/tidyseurat/actions"><img src="https://github.com/stemangiola/tidyseurat/workflows/R-CMD-check-bioc/badge.svg" alt="R build status"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
