<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Brings Seurat to the Tidyverse  • tidyseurat</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Brings Seurat to the Tidyverse ">
<meta property="og:description" content="It creates an invisible layer that allow to see the Seurat object 
    as tibble and interact seamlessly with the tidyverse.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">tidyseurat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.18</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/figures_article.html">Code for producing the figures in the article</a>
    </li>
    <li>
      <a href="articles/introduction.html">Overview of the tidyseurat package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div id="tidyseurat---part-of-tidytranscriptomics" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#tidyseurat---part-of-tidytranscriptomics" class="anchor"></a>tidyseurat - part of tidytranscriptomics</h1></div>
<!-- badges: start -->

<p><strong>Brings Seurat to the tidyverse!</strong></p>
<p>website: <a href="https://stemangiola.github.io/tidyseurat/">stemangiola.github.io/tidyseurat/</a></p>
<p>Please also have a look at</p>
<ul>
<li>
<a href="https://stemangiola.github.io/tidybulk/">tidybulk</a> for tidy bulk RNA-seq analysis</li>
<li>
<a href="https://github.com/stemangiola/nanny/">nanny</a> for tidy high-level data analysis and manipulation</li>
<li>
<a href="https://github.com/stemangiola/tidygate/">tidygate</a> for adding custom gate information to your tibble</li>
<li>
<a href="https://stemangiola.github.io/tidyHeatmap/">tidyHeatmap</a> for heatmaps produced with tidy principles</li>
</ul>
<p><img src="reference/figures/logo_interaction-01.png" alt="visual cue"></p>
</div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>tidyseurat provides a bridge between the Seurat single-cell package <br><span class="math display">@<em>b</em><em>u</em><em>t</em><em>l</em><em>e</em><em>r</em>2018<em>i</em><em>n</em><em>t</em><em>e</em><em>g</em><em>r</em><em>a</em><em>t</em><em>i</em><em>n</em><em>g</em>; @<em>s</em><em>t</em><em>u</em><em>a</em><em>r</em><em>t</em>2019<em>c</em><em>o</em><em>m</em><em>p</em><em>r</em><em>e</em><em>h</em><em>e</em><em>n</em><em>s</em><em>i</em><em>v</em><em>e</em></span><br> and the tidyverse <br><span class="math display">@<em>w</em><em>i</em><em>c</em><em>k</em><em>h</em><em>a</em><em>m</em>2019<em>w</em><em>e</em><em>l</em><em>c</em><em>o</em><em>m</em><em>e</em></span><br>. It creates an invisible layer that enables viewing the Seurat object as a tidyverse tibble, and provides Seurat-compatible <em>dplyr</em>, <em>tidyr</em>, <em>ggplot</em> and <em>plotly</em> functions.</p>
<div id="functionsutilities-available" class="section level2">
<h2 class="hasAnchor">
<a href="#functionsutilities-available" class="anchor"></a>Functions/utilities available</h2>
<table class="table">
<thead><tr class="header">
<th>Seurat-compatible Functions</th>
<th>Description</th>
</tr></thead>
<tbody><tr class="odd">
<td><code>all</code></td>
<td>After all <code>tidyseurat</code> is a Seurat object, just better</td>
</tr></tbody>
</table>
<table class="table">
<thead><tr class="header">
<th>tidyverse Packages</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>dplyr</code></td>
<td>All <code>dplyr</code> APIs like for any tibble</td>
</tr>
<tr class="even">
<td><code>tidyr</code></td>
<td>All <code>tidyr</code> APIs like for any tibble</td>
</tr>
<tr class="odd">
<td><code>ggplot2</code></td>
<td>
<code>ggplot</code> like for any tibble</td>
</tr>
<tr class="even">
<td><code>plotly</code></td>
<td>
<code>plot_ly</code> like for any tibble</td>
</tr>
</tbody>
</table>
<table class="table">
<thead><tr class="header">
<th>Utilities</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>tidy</code></td>
<td>Add <code>tidyseurat</code> invisible layer over a Seurat object</td>
</tr>
<tr class="even">
<td><code>as_tibble</code></td>
<td>Convert cell-wise information to a <code>tbl_df</code>
</td>
</tr>
<tr class="odd">
<td><code>join_transcripts</code></td>
<td>Add transcript-wise information, returns a <code>tbl_df</code>
</td>
</tr>
</tbody>
</table>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>From CRAN</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"tidyseurat"</span><span class="op">)</span></code></pre></div>
<p>From Github (development)</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"stemangiola/tidyseurat"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://purrr.tidyverse.org">purrr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org">magrittr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">tidyseurat</span><span class="op">)</span></code></pre></div>
</div>
<div id="create-tidyseurat-the-best-of-both-worlds" class="section level2">
<h2 class="hasAnchor">
<a href="#create-tidyseurat-the-best-of-both-worlds" class="anchor"></a>Create <code>tidyseurat</code>, the best of both worlds!</h2>
<p>This is a seurat object but it is evaluated as tibble. So it is fully compatible both with Seurat and tidyverse APIs.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_small_tidy</span> <span class="op">&lt;-</span> <span class="fu">tidyseurat</span><span class="fu">::</span><span class="va"><a href="reference/pbmc_small.html">pbmc_small</a></span> <span class="op">%&gt;%</span> <span class="fu"><a href="reference/tidy.html">tidy</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><strong>It looks like a tibble</strong></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_small_tidy</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="co">## # A tibble abstraction: 80 x 16</span></span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="co">##    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups</span></span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="co">##    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; </span></span>
<span id="cb6-4"><a href="#cb6-4"></a><span class="co">##  1 ATGC… SeuratPro…         70           47 0               A             g2    </span></span>
<span id="cb6-5"><a href="#cb6-5"></a><span class="co">##  2 CATG… SeuratPro…         85           52 0               A             g1    </span></span>
<span id="cb6-6"><a href="#cb6-6"></a><span class="co">##  3 GAAC… SeuratPro…         87           50 1               B             g2    </span></span>
<span id="cb6-7"><a href="#cb6-7"></a><span class="co">##  4 TGAC… SeuratPro…        127           56 0               A             g2    </span></span>
<span id="cb6-8"><a href="#cb6-8"></a><span class="co">##  5 AGTC… SeuratPro…        173           53 0               A             g2    </span></span>
<span id="cb6-9"><a href="#cb6-9"></a><span class="co">##  6 TCTG… SeuratPro…         70           48 0               A             g1    </span></span>
<span id="cb6-10"><a href="#cb6-10"></a><span class="co">##  7 TGGT… SeuratPro…         64           36 0               A             g1    </span></span>
<span id="cb6-11"><a href="#cb6-11"></a><span class="co">##  8 GCAG… SeuratPro…         72           45 0               A             g1    </span></span>
<span id="cb6-12"><a href="#cb6-12"></a><span class="co">##  9 GATA… SeuratPro…         52           36 0               A             g1    </span></span>
<span id="cb6-13"><a href="#cb6-13"></a><span class="co">## 10 AATG… SeuratPro…        100           41 0               A             g1    </span></span>
<span id="cb6-14"><a href="#cb6-14"></a><span class="co">## # … with 70 more rows, and 9 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,</span></span>
<span id="cb6-15"><a href="#cb6-15"></a><span class="co">## #   PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;,</span></span>
<span id="cb6-16"><a href="#cb6-16"></a><span class="co">## #   tSNE_2 &lt;dbl&gt;</span></span></code></pre></div>
<p><strong>But it is a Seurat object after all</strong></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_small_tidy</span><span class="op">@</span><span class="va">assays</span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="co">## $RNA</span></span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="co">## Assay data with 230 features for 80 cells</span></span>
<span id="cb8-3"><a href="#cb8-3"></a><span class="co">## Top 10 variable features:</span></span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="co">##  PPBP, IGLL5, VDAC3, CD1C, AKR1C3, PF4, MYL9, GNLY, TREML1, CA2</span></span></code></pre></div>
</div>
</div>
<div id="annotation-polishing" class="section level1">
<h1 class="hasAnchor">
<a href="#annotation-polishing" class="anchor"></a>Annotation polishing</h1>
<p>We may have a column that contains the directory each run was taken from, such as the “file” column in <code>pbmc_small_tidy</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_small_tidy</span><span class="op">$</span><span class="va">file</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="co">##                                     ATGCCAGAACGACT </span></span>
<span id="cb10-2"><a href="#cb10-2"></a><span class="co">## "../data/sample2/outs/filtered_feature_bc_matrix/" </span></span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="co">##                                     CATGGCCTGTGCAT </span></span>
<span id="cb10-4"><a href="#cb10-4"></a><span class="co">## "../data/sample1/outs/filtered_feature_bc_matrix/" </span></span>
<span id="cb10-5"><a href="#cb10-5"></a><span class="co">##                                     GAACCTGATGAACC </span></span>
<span id="cb10-6"><a href="#cb10-6"></a><span class="co">## "../data/sample2/outs/filtered_feature_bc_matrix/" </span></span>
<span id="cb10-7"><a href="#cb10-7"></a><span class="co">##                                     TGACTGGATTCTCA </span></span>
<span id="cb10-8"><a href="#cb10-8"></a><span class="co">## "../data/sample2/outs/filtered_feature_bc_matrix/" </span></span>
<span id="cb10-9"><a href="#cb10-9"></a><span class="co">##                                     AGTCAGACTGCACA </span></span>
<span id="cb10-10"><a href="#cb10-10"></a><span class="co">## "../data/sample2/outs/filtered_feature_bc_matrix/"</span></span></code></pre></div>
<p>We may want to extract the run/sample name out of it into a separate column. Tidyverse <code>extract</code> can be used to convert a character column into multiple columns using regular expression groups.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create sample column</span>
<span class="va">pbmc_small_polished</span> <span class="op">&lt;-</span>
  <span class="va">pbmc_small_tidy</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/unnest.html">extract</a></span><span class="op">(</span><span class="va">file</span>, <span class="st">"sample"</span>, <span class="st">"../data/([a-z0-9]+)/outs.+"</span>, remove <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="co"># Reorder to have sample column up front</span>
<span class="va">pbmc_small_polished</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/arrange.html">select</a></span><span class="op">(</span><span class="va">sample</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="co">## # A tibble abstraction: 80 x 17</span></span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="co">##    cell  sample orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents</span></span>
<span id="cb12-3"><a href="#cb12-3"></a><span class="co">##    &lt;chr&gt; &lt;chr&gt;  &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;        </span></span>
<span id="cb12-4"><a href="#cb12-4"></a><span class="co">##  1 ATGC… sampl… SeuratPro…         70           47 0               A            </span></span>
<span id="cb12-5"><a href="#cb12-5"></a><span class="co">##  2 CATG… sampl… SeuratPro…         85           52 0               A            </span></span>
<span id="cb12-6"><a href="#cb12-6"></a><span class="co">##  3 GAAC… sampl… SeuratPro…         87           50 1               B            </span></span>
<span id="cb12-7"><a href="#cb12-7"></a><span class="co">##  4 TGAC… sampl… SeuratPro…        127           56 0               A            </span></span>
<span id="cb12-8"><a href="#cb12-8"></a><span class="co">##  5 AGTC… sampl… SeuratPro…        173           53 0               A            </span></span>
<span id="cb12-9"><a href="#cb12-9"></a><span class="co">##  6 TCTG… sampl… SeuratPro…         70           48 0               A            </span></span>
<span id="cb12-10"><a href="#cb12-10"></a><span class="co">##  7 TGGT… sampl… SeuratPro…         64           36 0               A            </span></span>
<span id="cb12-11"><a href="#cb12-11"></a><span class="co">##  8 GCAG… sampl… SeuratPro…         72           45 0               A            </span></span>
<span id="cb12-12"><a href="#cb12-12"></a><span class="co">##  9 GATA… sampl… SeuratPro…         52           36 0               A            </span></span>
<span id="cb12-13"><a href="#cb12-13"></a><span class="co">## 10 AATG… sampl… SeuratPro…        100           41 0               A            </span></span>
<span id="cb12-14"><a href="#cb12-14"></a><span class="co">## # … with 70 more rows, and 10 more variables: groups &lt;chr&gt;,</span></span>
<span id="cb12-15"><a href="#cb12-15"></a><span class="co">## #   RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;,</span></span>
<span id="cb12-16"><a href="#cb12-16"></a><span class="co">## #   PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</span></span></code></pre></div>
</div>
<div id="preliminary-plots" class="section level1">
<h1 class="hasAnchor">
<a href="#preliminary-plots" class="anchor"></a>Preliminary plots</h1>
<p>Set colours and theme for plots.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Use colourblind-friendly colours</span>
<span class="va">friendly_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#88CCEE"</span>, <span class="st">"#CC6677"</span>, <span class="st">"#DDCC77"</span>, <span class="st">"#117733"</span>, <span class="st">"#332288"</span>, <span class="st">"#AA4499"</span>, <span class="st">"#44AA99"</span>, <span class="st">"#999933"</span>, <span class="st">"#882255"</span>, <span class="st">"#661100"</span>, <span class="st">"#6699CC"</span><span class="op">)</span>

<span class="co"># Set theme</span>
<span class="va">my_theme</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">friendly_cols</span><span class="op">)</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">friendly_cols</span><span class="op">)</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>
        panel.border <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
        axis.line <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span><span class="op">)</span>,
        panel.grid.major <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>,
        panel.grid.minor <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>,
        text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,
        legend.position <span class="op">=</span> <span class="st">"bottom"</span>,
        aspect.ratio <span class="op">=</span> <span class="fl">1</span>,
        strip.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
        axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span>t <span class="op">=</span> <span class="fl">10</span>, r <span class="op">=</span> <span class="fl">10</span>, b <span class="op">=</span> <span class="fl">10</span>, l <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span>,
        axis.title.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span>t <span class="op">=</span> <span class="fl">10</span>, r <span class="op">=</span> <span class="fl">10</span>, b <span class="op">=</span> <span class="fl">10</span>, l <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span>
      <span class="op">)</span>
  <span class="op">)</span></code></pre></div>
<p>We can treat <code>pbmc_small_polished</code> effectively as a normal tibble for plotting.</p>
<p>Here we plot number of transcripts per cell.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_small_polished</span> <span class="op">%&gt;%</span>
  <span class="fu">tidyseurat</span><span class="fu">::</span><span class="fu"><a href="reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">nFeature_RNA</span>, fill <span class="op">=</span> <span class="va">groups</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">my_theme</span></code></pre></div>
<p><img src="reference/figures/plot1-1.png"><!-- --></p>
<p>Here we plot total transcripts per cell.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_small_polished</span> <span class="op">%&gt;%</span>
  <span class="fu">tidyseurat</span><span class="fu">::</span><span class="fu"><a href="reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">groups</span>, <span class="va">nCount_RNA</span>, fill <span class="op">=</span> <span class="va">groups</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span>outlier.shape <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">my_theme</span></code></pre></div>
<p><img src="reference/figures/plot2-1.png"><!-- --></p>
<p>Here we plot abundance of two transcripts for each group.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_small_polished</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/join_transcripts.html">join_transcripts</a></span><span class="op">(</span>transcripts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"HLA-DRA"</span>, <span class="st">"LYZ"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">groups</span>, <span class="va">abundance_RNA</span> <span class="op">+</span> <span class="fl">1</span>, fill <span class="op">=</span> <span class="va">groups</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span>outlier.shape <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>size <span class="op">=</span> <span class="va">nCount_RNA</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.5</span>, width <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">my_theme</span></code></pre></div>
<p><img src="reference/figures/unnamed-chunk-14-1.png"><!-- --></p>
</div>
<div id="preprocess-the-dataset" class="section level1">
<h1 class="hasAnchor">
<a href="#preprocess-the-dataset" class="anchor"></a>Preprocess the dataset</h1>
<p>Also you can treat the object as Seurat object and proceed with data processing.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_small_pca</span> <span class="op">&lt;-</span>
  <span class="va">pbmc_small_polished</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html">SCTransform</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html">FindVariableFeatures</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="va">pbmc_small_pca</span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="co">## # A tibble abstraction: 80 x 19</span></span>
<span id="cb18-2"><a href="#cb18-2"></a><span class="co">##    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups</span></span>
<span id="cb18-3"><a href="#cb18-3"></a><span class="co">##    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; </span></span>
<span id="cb18-4"><a href="#cb18-4"></a><span class="co">##  1 ATGC… SeuratPro…         70           47 0               A             g2    </span></span>
<span id="cb18-5"><a href="#cb18-5"></a><span class="co">##  2 CATG… SeuratPro…         85           52 0               A             g1    </span></span>
<span id="cb18-6"><a href="#cb18-6"></a><span class="co">##  3 GAAC… SeuratPro…         87           50 1               B             g2    </span></span>
<span id="cb18-7"><a href="#cb18-7"></a><span class="co">##  4 TGAC… SeuratPro…        127           56 0               A             g2    </span></span>
<span id="cb18-8"><a href="#cb18-8"></a><span class="co">##  5 AGTC… SeuratPro…        173           53 0               A             g2    </span></span>
<span id="cb18-9"><a href="#cb18-9"></a><span class="co">##  6 TCTG… SeuratPro…         70           48 0               A             g1    </span></span>
<span id="cb18-10"><a href="#cb18-10"></a><span class="co">##  7 TGGT… SeuratPro…         64           36 0               A             g1    </span></span>
<span id="cb18-11"><a href="#cb18-11"></a><span class="co">##  8 GCAG… SeuratPro…         72           45 0               A             g1    </span></span>
<span id="cb18-12"><a href="#cb18-12"></a><span class="co">##  9 GATA… SeuratPro…         52           36 0               A             g1    </span></span>
<span id="cb18-13"><a href="#cb18-13"></a><span class="co">## 10 AATG… SeuratPro…        100           41 0               A             g1    </span></span>
<span id="cb18-14"><a href="#cb18-14"></a><span class="co">## # … with 70 more rows, and 12 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,</span></span>
<span id="cb18-15"><a href="#cb18-15"></a><span class="co">## #   sample &lt;chr&gt;, nCount_SCT &lt;dbl&gt;, nFeature_SCT &lt;int&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;,</span></span>
<span id="cb18-16"><a href="#cb18-16"></a><span class="co">## #   PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</span></span></code></pre></div>
<p>If a tool is not included in the tidyseurat collection, we can use <code>as_tibble</code> to permanently convert <code>tidyseurat</code> into tibble.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_small_pca</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/arrange.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">contains</a></span><span class="op">(</span><span class="st">"PC"</span><span class="op">)</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">GGally</span><span class="fu">::</span><span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html">ggpairs</a></span><span class="op">(</span>columns <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">groups</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">my_theme</span></code></pre></div>
<p><img src="reference/figures/pc_plot-1.png"><!-- --></p>
</div>
<div id="identify-clusters" class="section level1">
<h1 class="hasAnchor">
<a href="#identify-clusters" class="anchor"></a>Identify clusters</h1>
<p>We proceed with cluster identification with Seurat.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_small_cluster</span> <span class="op">&lt;-</span>
  <span class="va">pbmc_small_pca</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"igraph"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="va">pbmc_small_cluster</span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="co">## # A tibble abstraction: 80 x 21</span></span>
<span id="cb21-2"><a href="#cb21-2"></a><span class="co">##    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups</span></span>
<span id="cb21-3"><a href="#cb21-3"></a><span class="co">##    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; </span></span>
<span id="cb21-4"><a href="#cb21-4"></a><span class="co">##  1 ATGC… SeuratPro…         70           47 0               A             g2    </span></span>
<span id="cb21-5"><a href="#cb21-5"></a><span class="co">##  2 CATG… SeuratPro…         85           52 0               A             g1    </span></span>
<span id="cb21-6"><a href="#cb21-6"></a><span class="co">##  3 GAAC… SeuratPro…         87           50 1               B             g2    </span></span>
<span id="cb21-7"><a href="#cb21-7"></a><span class="co">##  4 TGAC… SeuratPro…        127           56 0               A             g2    </span></span>
<span id="cb21-8"><a href="#cb21-8"></a><span class="co">##  5 AGTC… SeuratPro…        173           53 0               A             g2    </span></span>
<span id="cb21-9"><a href="#cb21-9"></a><span class="co">##  6 TCTG… SeuratPro…         70           48 0               A             g1    </span></span>
<span id="cb21-10"><a href="#cb21-10"></a><span class="co">##  7 TGGT… SeuratPro…         64           36 0               A             g1    </span></span>
<span id="cb21-11"><a href="#cb21-11"></a><span class="co">##  8 GCAG… SeuratPro…         72           45 0               A             g1    </span></span>
<span id="cb21-12"><a href="#cb21-12"></a><span class="co">##  9 GATA… SeuratPro…         52           36 0               A             g1    </span></span>
<span id="cb21-13"><a href="#cb21-13"></a><span class="co">## 10 AATG… SeuratPro…        100           41 0               A             g1    </span></span>
<span id="cb21-14"><a href="#cb21-14"></a><span class="co">## # … with 70 more rows, and 14 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,</span></span>
<span id="cb21-15"><a href="#cb21-15"></a><span class="co">## #   sample &lt;chr&gt;, nCount_SCT &lt;dbl&gt;, nFeature_SCT &lt;int&gt;, SCT_snn_res.0.8 &lt;fct&gt;,</span></span>
<span id="cb21-16"><a href="#cb21-16"></a><span class="co">## #   seurat_clusters &lt;fct&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;,</span></span>
<span id="cb21-17"><a href="#cb21-17"></a><span class="co">## #   PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</span></span></code></pre></div>
<p>Now we can interrogate the object as if it was a regular tibble data frame.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_small_cluster</span> <span class="op">%&gt;%</span>
  <span class="fu">tidyseurat</span><span class="fu">::</span><span class="fu"><a href="reference/count.html">count</a></span><span class="op">(</span><span class="va">groups</span>, <span class="va">seurat_clusters</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a><span class="co">## # A tibble: 8 x 3</span></span>
<span id="cb23-2"><a href="#cb23-2"></a><span class="co">##   groups seurat_clusters     n</span></span>
<span id="cb23-3"><a href="#cb23-3"></a><span class="co">##   &lt;chr&gt;  &lt;fct&gt;           &lt;int&gt;</span></span>
<span id="cb23-4"><a href="#cb23-4"></a><span class="co">## 1 g1     0                  17</span></span>
<span id="cb23-5"><a href="#cb23-5"></a><span class="co">## 2 g1     1                  14</span></span>
<span id="cb23-6"><a href="#cb23-6"></a><span class="co">## 3 g1     2                   9</span></span>
<span id="cb23-7"><a href="#cb23-7"></a><span class="co">## 4 g1     3                   4</span></span>
<span id="cb23-8"><a href="#cb23-8"></a><span class="co">## 5 g2     0                  13</span></span>
<span id="cb23-9"><a href="#cb23-9"></a><span class="co">## 6 g2     1                  12</span></span>
<span id="cb23-10"><a href="#cb23-10"></a><span class="co">## 7 g2     2                   6</span></span>
<span id="cb23-11"><a href="#cb23-11"></a><span class="co">## 8 g2     3                   5</span></span></code></pre></div>
<p>We can identify cluster markers using Seurat.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Identify top 10 markers per cluster</span>
<span class="va">markers</span> <span class="op">&lt;-</span>
  <span class="va">pbmc_small_cluster</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers</a></span><span class="op">(</span>only.pos <span class="op">=</span> <span class="cn">TRUE</span>, min.pct <span class="op">=</span> <span class="fl">0.25</span>, thresh.use <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/arrange.html">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html">top_n</a></span><span class="op">(</span><span class="fl">10</span>, <span class="va">avg_logFC</span><span class="op">)</span>

<span class="co"># Plot heatmap</span>
<span class="va">pbmc_small_cluster</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://satijalab.org/seurat/reference/DoHeatmap.html">DoHeatmap</a></span><span class="op">(</span>
    features <span class="op">=</span> <span class="va">markers</span><span class="op">$</span><span class="va">gene</span>,
    group.colors <span class="op">=</span> <span class="va">friendly_cols</span>
  <span class="op">)</span></code></pre></div>
<p><img src="reference/figures/unnamed-chunk-15-1.png"><!-- --></p>
</div>
<div id="reduce-dimensions" class="section level1">
<h1 class="hasAnchor">
<a href="#reduce-dimensions" class="anchor"></a>Reduce dimensions</h1>
<p>We can calculate the first 3 UMAP dimensions using the Seurat framework.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_small_UMAP</span> <span class="op">&lt;-</span>
  <span class="va">pbmc_small_cluster</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>, n.components <span class="op">=</span> <span class="fl">3L</span>, <span class="op">)</span></code></pre></div>
<p>And we can plot them using 3D plot using plotly.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_small_UMAP</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/plot_ly.html">plot_ly</a></span><span class="op">(</span>
    x <span class="op">=</span> <span class="op">~</span><span class="va">`UMAP_1`</span>,
    y <span class="op">=</span> <span class="op">~</span><span class="va">`UMAP_2`</span>,
    z <span class="op">=</span> <span class="op">~</span><span class="va">`UMAP_3`</span>,
    color <span class="op">=</span> <span class="op">~</span><span class="va">seurat_clusters</span>,
    colors <span class="op">=</span> <span class="va">friendly_cols</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>
  <span class="op">)</span></code></pre></div>
<p><img src="reference/figures/plotly.png" alt="screenshot plotly"></p>
<div id="cell-type-prediction" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-type-prediction" class="anchor"></a>Cell type prediction</h2>
<p>We can infer cell type identities using <em>SingleR</em> <br><span class="math display">@<em>a</em><em>r</em><em>a</em><em>n</em>2019<em>r</em><em>e</em><em>f</em><em>e</em><em>r</em><em>e</em><em>n</em><em>c</em><em>e</em></span><br> and manipulate the output using tidyverse.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Get cell type reference data</span>
<span class="va">blueprint</span> <span class="op">&lt;-</span> <span class="fu">celldex</span><span class="fu">::</span><span class="fu">BlueprintEncodeData</span><span class="op">(</span><span class="op">)</span>

<span class="co"># Infer cell identities</span>
<span class="va">cell_type_df</span> <span class="op">&lt;-</span>
  <span class="va">pbmc_small_UMAP</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"SCT"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">counts</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log1p</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/Matrix.html">Matrix</a></span><span class="op">(</span>sparse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">SingleR</span><span class="fu">::</span><span class="fu">SingleR</span><span class="op">(</span>
    ref <span class="op">=</span> <span class="va">blueprint</span>,
    labels <span class="op">=</span> <span class="va">blueprint</span><span class="op">$</span><span class="va">label.main</span>,
    method <span class="op">=</span> <span class="st">"single"</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/as_tibble.html">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">"cell"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/arrange.html">select</a></span><span class="op">(</span><span class="va">cell</span>, <span class="va">first.labels</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Join UMAP and cell type info</span>
<span class="va">pbmc_small_cell_type</span> <span class="op">&lt;-</span>
  <span class="va">pbmc_small_UMAP</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/arrange.html">left_join</a></span><span class="op">(</span><span class="va">cell_type_df</span>, by <span class="op">=</span> <span class="st">"cell"</span><span class="op">)</span>

<span class="co"># Reorder columns</span>
<span class="va">pbmc_small_cell_type</span> <span class="op">%&gt;%</span>
  <span class="fu">tidyseurat</span><span class="fu">::</span><span class="fu"><a href="reference/arrange.html">select</a></span><span class="op">(</span><span class="va">cell</span>, <span class="va">first.labels</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb29"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a><span class="co">## # A tibble abstraction: 80 x 25</span></span>
<span id="cb29-2"><a href="#cb29-2"></a><span class="co">##    cell  first.labels orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8</span></span>
<span id="cb29-3"><a href="#cb29-3"></a><span class="co">##    &lt;chr&gt; &lt;chr&gt;        &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;          </span></span>
<span id="cb29-4"><a href="#cb29-4"></a><span class="co">##  1 ATGC… CD4+ T-cells SeuratPro…         70           47 0              </span></span>
<span id="cb29-5"><a href="#cb29-5"></a><span class="co">##  2 CATG… CD8+ T-cells SeuratPro…         85           52 0              </span></span>
<span id="cb29-6"><a href="#cb29-6"></a><span class="co">##  3 GAAC… CD8+ T-cells SeuratPro…         87           50 1              </span></span>
<span id="cb29-7"><a href="#cb29-7"></a><span class="co">##  4 TGAC… CD4+ T-cells SeuratPro…        127           56 0              </span></span>
<span id="cb29-8"><a href="#cb29-8"></a><span class="co">##  5 AGTC… CD4+ T-cells SeuratPro…        173           53 0              </span></span>
<span id="cb29-9"><a href="#cb29-9"></a><span class="co">##  6 TCTG… CD4+ T-cells SeuratPro…         70           48 0              </span></span>
<span id="cb29-10"><a href="#cb29-10"></a><span class="co">##  7 TGGT… CD4+ T-cells SeuratPro…         64           36 0              </span></span>
<span id="cb29-11"><a href="#cb29-11"></a><span class="co">##  8 GCAG… CD4+ T-cells SeuratPro…         72           45 0              </span></span>
<span id="cb29-12"><a href="#cb29-12"></a><span class="co">##  9 GATA… CD4+ T-cells SeuratPro…         52           36 0              </span></span>
<span id="cb29-13"><a href="#cb29-13"></a><span class="co">## 10 AATG… CD4+ T-cells SeuratPro…        100           41 0              </span></span>
<span id="cb29-14"><a href="#cb29-14"></a><span class="co">## # … with 70 more rows, and 19 more variables: letter.idents &lt;fct&gt;,</span></span>
<span id="cb29-15"><a href="#cb29-15"></a><span class="co">## #   groups &lt;chr&gt;, RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;, sample &lt;chr&gt;,</span></span>
<span id="cb29-16"><a href="#cb29-16"></a><span class="co">## #   nCount_SCT &lt;dbl&gt;, nFeature_SCT &lt;int&gt;, SCT_snn_res.0.8 &lt;fct&gt;,</span></span>
<span id="cb29-17"><a href="#cb29-17"></a><span class="co">## #   seurat_clusters &lt;fct&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;,</span></span>
<span id="cb29-18"><a href="#cb29-18"></a><span class="co">## #   PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;, UMAP_1 &lt;dbl&gt;, UMAP_2 &lt;dbl&gt;,</span></span>
<span id="cb29-19"><a href="#cb29-19"></a><span class="co">## #   UMAP_3 &lt;dbl&gt;</span></span></code></pre></div>
<p>We can easily summarise the results. For example, we can see how cell type classification overlaps with cluster classification.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_small_cell_type</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/count.html">count</a></span><span class="op">(</span><span class="va">seurat_clusters</span>, <span class="va">first.labels</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb31"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a><span class="co">## # A tibble: 9 x 3</span></span>
<span id="cb31-2"><a href="#cb31-2"></a><span class="co">##   seurat_clusters first.labels     n</span></span>
<span id="cb31-3"><a href="#cb31-3"></a><span class="co">##   &lt;fct&gt;           &lt;chr&gt;        &lt;int&gt;</span></span>
<span id="cb31-4"><a href="#cb31-4"></a><span class="co">## 1 0               CD4+ T-cells     8</span></span>
<span id="cb31-5"><a href="#cb31-5"></a><span class="co">## 2 0               CD8+ T-cells    10</span></span>
<span id="cb31-6"><a href="#cb31-6"></a><span class="co">## 3 0               NK cells        12</span></span>
<span id="cb31-7"><a href="#cb31-7"></a><span class="co">## 4 1               Macrophages      1</span></span>
<span id="cb31-8"><a href="#cb31-8"></a><span class="co">## 5 1               Monocytes       25</span></span>
<span id="cb31-9"><a href="#cb31-9"></a><span class="co">## 6 2               B-cells         10</span></span>
<span id="cb31-10"><a href="#cb31-10"></a><span class="co">## 7 2               Macrophages      1</span></span>
<span id="cb31-11"><a href="#cb31-11"></a><span class="co">## 8 2               Monocytes        4</span></span>
<span id="cb31-12"><a href="#cb31-12"></a><span class="co">## 9 3               Erythrocytes     9</span></span></code></pre></div>
<p>We can easily reshape the data for building information-rich faceted plots.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_small_cell_type</span> <span class="op">%&gt;%</span>

  <span class="co"># Reshape and add classifier column</span>
  <span class="fu"><a href="reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>
    cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">seurat_clusters</span>, <span class="va">first.labels</span><span class="op">)</span>,
    names_to <span class="op">=</span> <span class="st">"classifier"</span>, values_to <span class="op">=</span> <span class="st">"label"</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># UMAP plots for cell type and cluster</span>
  <span class="fu"><a href="reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">UMAP_1</span>, <span class="va">UMAP_2</span>, color <span class="op">=</span> <span class="va">label</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">classifier</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">my_theme</span></code></pre></div>
<p><img src="reference/figures/unnamed-chunk-19-1.png"><!-- --></p>
<p>We can easily plot gene correlation per cell category, adding multi-layer annotations.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_small_cell_type</span> <span class="op">%&gt;%</span>

  <span class="co"># Add some mitochondrial abundance values</span>
  <span class="fu"><a href="reference/arrange.html">mutate</a></span><span class="op">(</span>mitochondrial <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Plot correlation</span>
  <span class="fu"><a href="reference/join_transcripts.html">join_transcripts</a></span><span class="op">(</span>transcripts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CST3"</span>, <span class="st">"LYZ"</span><span class="op">)</span>, shape <span class="op">=</span> <span class="st">"wide"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">CST3</span> <span class="op">+</span> <span class="fl">1</span>, <span class="va">LYZ</span> <span class="op">+</span> <span class="fl">1</span>, color <span class="op">=</span> <span class="va">groups</span>, size <span class="op">=</span> <span class="va">mitochondrial</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">first.labels</span>, scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">my_theme</span></code></pre></div>
<p><img src="reference/figures/unnamed-chunk-20-1.png"><!-- --></p>
</div>
</div>
<div id="nested-analyses" class="section level1">
<h1 class="hasAnchor">
<a href="#nested-analyses" class="anchor"></a>Nested analyses</h1>
<p>A powerful tool we can use with tidyseurat is <code>nest</code>. We can easily perform independent analyses on subsets of the dataset. First we classify cell types in lymphoid and myeloid; then, nest based on the new classification</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_small_nested</span> <span class="op">&lt;-</span>
  <span class="va">pbmc_small_cell_type</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/arrange.html">filter</a></span><span class="op">(</span><span class="va">first.labels</span> <span class="op">!=</span> <span class="st">"Erythrocytes"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/arrange.html">mutate</a></span><span class="op">(</span>cell_class <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="va">`first.labels`</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Macrophages"</span>, <span class="st">"Monocytes"</span><span class="op">)</span>, <span class="st">"myeloid"</span>, <span class="st">"lymphoid"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/unnest.html">nest</a></span><span class="op">(</span>data <span class="op">=</span> <span class="op">-</span><span class="va">cell_class</span><span class="op">)</span>

<span class="va">pbmc_small_nested</span></code></pre></div>
<div class="sourceCode" id="cb35"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a><span class="co">## # A tibble: 2 x 2</span></span>
<span id="cb35-2"><a href="#cb35-2"></a><span class="co">##   cell_class data      </span></span>
<span id="cb35-3"><a href="#cb35-3"></a><span class="co">##   &lt;chr&gt;      &lt;list&gt;    </span></span>
<span id="cb35-4"><a href="#cb35-4"></a><span class="co">## 1 lymphoid   &lt;tidysert&gt;</span></span>
<span id="cb35-5"><a href="#cb35-5"></a><span class="co">## 2 myeloid    &lt;tidysert&gt;</span></span></code></pre></div>
<p>Now we can independently for the lymphoid and myeloid subsets (i) find variable features, (ii) reduce dimensions, and (iii) cluster using both tidyverse and SingleCellExperiment seamlessly.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_small_nested_reanalysed</span> <span class="op">&lt;-</span>
  <span class="va">pbmc_small_nested</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/arrange.html">mutate</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span>
    <span class="va">data</span>, <span class="op">~</span> <span class="va">.x</span> <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html">FindVariableFeatures</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span>npcs <span class="op">=</span> <span class="fl">10</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"igraph"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, n.components <span class="op">=</span> <span class="fl">3L</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
  <span class="op">)</span><span class="op">)</span>

<span class="va">pbmc_small_nested_reanalysed</span></code></pre></div>
<div class="sourceCode" id="cb37"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1"></a><span class="co">## # A tibble: 2 x 2</span></span>
<span id="cb37-2"><a href="#cb37-2"></a><span class="co">##   cell_class data      </span></span>
<span id="cb37-3"><a href="#cb37-3"></a><span class="co">##   &lt;chr&gt;      &lt;list&gt;    </span></span>
<span id="cb37-4"><a href="#cb37-4"></a><span class="co">## 1 lymphoid   &lt;tidysert&gt;</span></span>
<span id="cb37-5"><a href="#cb37-5"></a><span class="co">## 2 myeloid    &lt;tidysert&gt;</span></span></code></pre></div>
<p>Now we can unnest and plot the new classification.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_small_nested_reanalysed</span> <span class="op">%&gt;%</span>

  <span class="co"># Convert to tibble otherwise Seurat drops reduced dimensions when unifying data sets.</span>
  <span class="fu"><a href="reference/arrange.html">mutate</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="va">data</span>, <span class="op">~</span> <span class="va">.x</span> <span class="op">%&gt;%</span> <span class="fu"><a href="reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/unnest.html">unnest</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Define unique clusters</span>
  <span class="fu"><a href="reference/unnest.html">unite</a></span><span class="op">(</span><span class="st">"cluster"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">cell_class</span>, <span class="va">seurat_clusters</span><span class="op">)</span>, remove <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Plotting</span>
  <span class="fu"><a href="reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">UMAP_1</span>, <span class="va">UMAP_2</span>, color <span class="op">=</span> <span class="va">cluster</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">cell_class</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">my_theme</span></code></pre></div>
<p><img src="reference/figures/unnamed-chunk-23-1.png"><!-- --></p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from CRAN at <br><a href="https://cloud.r-project.org/package=tidyseurat">https://​cloud.r-project.org/​package=tidyseurat</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Stefano Mangiola <br><small class="roles"> Author, maintainer </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.tidyverse.org/lifecycle/#maturing"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a></li>
<li><a href="https://github.com/stemangiola/tidyseurat/actions/"><img src="https://github.com/stemangiola/tidyseurat/workflows/R-CMD-check/badge.svg" alt="R build status"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
